{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "71b5148d-57e8-44c9-b9b7-48af15f37eea",
   "metadata": {},
   "source": [
    "# Variant Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "931073ac-6333-4c18-bab1-bfd039ce381d",
   "metadata": {},
   "source": [
    "### Steps\n",
    "1. Remove monomorphic REF sites <br/>\n",
    "   `bcftools view -m2 -Oz -o out.vcf.gz in.vcf.gz`\n",
    "    <br/>\n",
    "   \n",
    "3. Remove INDELs (with SnpGap 0/5/10) <br/>\n",
    "   `bcftools filter --SnpGap <INT> in.vcf.gz | bcftools view -V indels -Oz -o out.vcf.gz`\n",
    "   <br/>\n",
    "\n",
    "5. Remove monomorphic SNPs <br/>\n",
    "   `bcftools view -e \"AC==AN || AC==0\" -Oz -o $outVCF $inVCF`\n",
    "   <br/>\n",
    "\n",
    "7. Remove multi-allelic SNPs <br/>\n",
    "   `bcftools view -m2 -M2 -Oz -o $outVCF $inVCF`\n",
    "   <br/>\n",
    "\n",
    "9. Filter based on DEPTH, QUALITY, and MQ. <br/>\n",
    "   `bcftools filter -e 'INFO/DP<500 | INFO/DP>7732 | QUAL<20 | MQ<30' -Oz -o $outVCF $inVCF`\n",
    "   <br/>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dff6e4d2-c82c-4dd9-afb0-c752faaa0180",
   "metadata": {},
   "source": [
    "__NB:__ \"SNPs\" in VCF files incl. all sites that are \n",
    "- bi-allelic SNPs, ie. only 1 ALT allele\n",
    "- multi-allelic SNPs ie. >1 ALT allele\n",
    "- fixed at ALT\n",
    "\n",
    "<br/>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fbab155d-d34b-42a5-a48d-d46ce372f70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load modules\n",
    "module load bcftools vcftools python3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d39c68a0-82b1-4937-980d-32ee7278aea3",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "## Test variant filtering steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e86b06c1-b231-4edb-a7b9-91fdd7c3cb9a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# This file was produced by bcftools stats (1.16+htslib-1.16) and can be plotted using plot-vcfstats.\n",
      "# The command line was:\tbcftools stats  --threads 10 -s- --af-bins /dev/fd/63 --depth 0,25000,25 /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all/test/test.vcf.gz\n",
      "#\n",
      "# Definition of sets:\n",
      "# ID\t[2]id\t[3]tab-separated file names\n",
      "ID\t0\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all/test/test.vcf.gz\n",
      "# SN, Summary numbers:\n",
      "#   number of records   .. number of data rows in the VCF\n",
      "#   number of no-ALTs   .. reference-only sites, ALT is either \".\" or identical to REF\n",
      "#   number of SNPs      .. number of rows with a SNP\n",
      "#   number of MNPs      .. number of rows with a MNP, such as CC>TT\n",
      "#   number of indels    .. number of rows with an indel\n",
      "#   number of others    .. number of rows with other type, for example a symbolic allele or\n",
      "#                          a complex substitution, such as ACT>TCGA\n",
      "#   number of multiallelic sites     .. number of rows with multiple alternate alleles\n",
      "#   number of multiallelic SNP sites .. number of rows with multiple alternate alleles, all SNPs\n",
      "# \n",
      "#   Note that rows containing multiple types will be counted multiple times, in each\n",
      "#   counter. For example, a row with a SNP and an indel increments both the SNP and\n",
      "#   the indel counter.\n",
      "# \n",
      "# SN\t[2]id\t[3]key\t[4]value\n",
      "SN\t0\tnumber of samples:\t1074\n",
      "SN\t0\tnumber of records:\t10085\n",
      "SN\t0\tnumber of no-ALTs:\t9720\n",
      "SN\t0\tnumber of SNPs:\t326\n",
      "SN\t0\tnumber of MNPs:\t0\n",
      "SN\t0\tnumber of indels:\t39\n",
      "SN\t0\tnumber of others:\t0\n",
      "SN\t0\tnumber of multiallelic sites:\t19\n",
      "SN\t0\tnumber of multiallelic SNP sites:\t16\n"
     ]
    }
   ],
   "source": [
    "cd   ~/snap_hap/variants/vcf_bcftools_Am_all/test\n",
    "\n",
    "## Test VCF\n",
    "bcftools view -r Chr8:1-10000 -Oz -o ~/snap_hap/variants/vcf_bcftools_Am_all/test/test.vcf.gz \\\n",
    "    ~/snap_hap/variants/vcf_bcftools_Am_all/Chr8/Am_all_bcftools_Chr8.vcf.gz\n",
    "\n",
    "## Initialize variables\n",
    "inVCF=/nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all/test/test.vcf.gz \n",
    "\n",
    "## Index VCF\n",
    "bcftools tabix -f $inVCF\n",
    "\n",
    "## bcftools stats\n",
    "bcftools stats --threads 10 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 $inVCF > ${inVCF/.vcf.gz/_stats.vchk}\n",
    "head -31 ${inVCF/.vcf.gz/_stats.vchk}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5e66178f-78ff-418e-a396-b91eba53b3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd   ~/snap_hap/variants/vcf_bcftools_Am_all/test\n",
    "\n",
    "## 1. Remove monomorphic REF sites\n",
    "inVCF=test.vcf.gz\n",
    "outVCF=test_allSNPs+allINDELs.vcf.gz\n",
    "bcftools view --threads 20 -m2 -Oz -o $outVCF $inVCF\n",
    "bcftools stats --threads 20 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 $outVCF > ${outVCF/.vcf.gz/_stats.vchk}\n",
    "\n",
    "\n",
    "## 2. Remove INDELs (with SnpGap 0)\n",
    "inVCF=test_allSNPs+allINDELs.vcf.gz\n",
    "outVCF=test_SnpGap0_allSNPs.vcf.gz\n",
    "bcftools view --threads 20 -V indels -Oz -o $outVCF $inVCF\n",
    "bcftools stats --threads 20 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 $outVCF > ${outVCF/.vcf.gz/_stats.vchk}\n",
    "\n",
    "\n",
    "## 3. Remove monomorphic SNPs\n",
    "inVCF=test_SnpGap0_allSNPs.vcf.gz\n",
    "outVCF=test_SnpGap0_polySNPs.vcf.gz\n",
    "bcftools view --threads 20 -e AC==AN || AC==0\" -Oz -o $outVCF $inVCF\n",
    "bcftools stats --threads 20 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 $outVCF > ${outVCF/.vcf.gz/_stats.vchk}\n",
    "\n",
    "\n",
    "## 4. Remove multi-allelic SNPs\n",
    "inVCF=test_SnpGap0_polySNPs.vcf.gz\n",
    "outVCF=test_SnpGap0_biSNPs.vcf.gz\n",
    "bcftools view --threads 20 -m2 -M2 -v snps -Oz -o $outVCF $inVCF\n",
    "bcftools stats --threads 20 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 $outVCF > ${outVCF/.vcf.gz/_stats.vchk}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5bd9a8c5-ac1f-48f9-b120-a7a8033fbc25",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Records from the original VCF\n",
      "Total records = 10085\n",
      "Fixed REFs = 9720\n",
      "without Fixed REFs = 365\n",
      "Fixed ALT indels = 25\n",
      "Fixed ALT snps = 16\n",
      "Multi-allelic sites = 19\n",
      "indels at multi-allelic sites = 3\n",
      "snps at multi-allelic sites = 16\n",
      "indels at TRUE bi-allelic sites = 11\n",
      "snps at TRUE biallelic sites = 294\n",
      "\n",
      "\n",
      "Records from the filtered VCF\n",
      "Total Records: 10085\n",
      "allSNPs + allINDELs: 365\n",
      "allSNPs: 326\n",
      "polySNPs: 310\n",
      "biSNPs: 294\n"
     ]
    }
   ],
   "source": [
    "echo -e \"Records from the original VCF\"\n",
    "\n",
    "echo Total records = `bcftools view -H test.vcf.gz | wc -l`\n",
    "echo Fixed REFs = `bcftools view -H -M1 test.vcf.gz | wc -l`\n",
    "echo without Fixed REFs = `bcftools view -H -m2 test.vcf.gz | wc -l`\n",
    "\n",
    "echo Fixed ALT indels = `bcftools view -H -v indels -i \"AC==AN\" test.vcf.gz | wc -l`\n",
    "echo Fixed ALT snps = `bcftools view -H -v snps -i \"AC==AN\" test.vcf.gz | wc -l`\n",
    "\n",
    "echo Multi-allelic sites = `bcftools view -H -m3 -e \"AC==AN\" test.vcf.gz | wc -l`\n",
    "echo indels at multi-allelic sites = `bcftools view -H -v indels -m3 -e \"AC==AN\" test.vcf.gz | wc -l`\n",
    "echo snps at multi-allelic sites = `bcftools view -H -v snps -m3 -e \"AC==AN\" test.vcf.gz | wc -l`\n",
    "\n",
    "echo indels at TRUE bi-allelic sites =  `bcftools view -H -v indels -m2 -M2 -e \"AC==AN\" test.vcf.gz | wc -l`\n",
    "echo snps at TRUE biallelic sites = `bcftools view -H -v snps -m2 -M2 -e \"AC==AN\" test.vcf.gz | wc -l`\n",
    "\n",
    "echo -e \"\\n\\nRecords from the filtered VCF\"\n",
    "echo Total Records: `bcftools view -H test.vcf.gz | wc -l` \n",
    "echo allSNPs + allINDELs: `bcftools view -H test_allSNPs+allINDELs.vcf.gz | wc -l`\n",
    "echo allSNPs: `bcftools view -H test_SnpGap0_allSNPs.vcf.gz | wc -l`\n",
    "echo polySNPs: `bcftools view -H test_SnpGap0_polySNPs.vcf.gz | wc -l`\n",
    "echo biSNPs: `bcftools view -H test_SnpGap0_biSNPs.vcf.gz | wc -l`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50836325-750e-4b48-8c93-10001d8fdf4e",
   "metadata": {},
   "source": [
    "<br/>\n",
    "<hr/>\n",
    "\n",
    "## Variant Filtering on cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c188258-c297-41e4-a08b-76eb69e5b222",
   "metadata": {},
   "source": [
    "### Filter1: Remove monomorphic REFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c3591a6-f5c5-4acb-82d3-613a054b0404",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/4-variant-filtering\n",
    "#chr=Chr1\n",
    "#threads=10\n",
    "\n",
    "for chr in Chr{1..8}\n",
    "do\n",
    "    sbatch -J ${chr}-filter1-removeREFs ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_filter1-removeREFs.sbatch \\\n",
    "        ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}.vcf.gz \\\n",
    "        ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_allSNPs+allINDELs.vcf.gz \\\n",
    "        $threads\n",
    "done\n",
    "\n",
    "## Run bcftools stats\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/3-stats\n",
    "for inVCF in ~/snap_hap/variants/vcf_bcftools_Am_all/*/Am_all_bcftools_*_allSNPs+allINDELs.vcf.gz\n",
    "do\n",
    "    echo $inVCF\n",
    "    sbatch -J vcf-stats ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_stats.sbatch $inVCF $threads\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ea8166b-12d0-4629-addb-eb4b35458eff",
   "metadata": {},
   "source": [
    "```\n",
    "#SBATCH --partition=defaultp\n",
    "#SBATCH --exclude=zeta[243-262],beta[231-235]\n",
    "#SBATCH --time=120:00:00\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem-per-cpu=8G\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --ntasks=1\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c91ea29d-c34e-41d8-be6f-29a90702956a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Filter2: Remove INDELs (incl. SNPs in vicinity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d42de6-e339-40ac-bc18-bca4d7455628",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/4-variant-filtering\n",
    "#chr=Chr1 ## Chr{1..8}\n",
    "#threads=10 \n",
    "#SnpGap=10 ## SnpGap: 0,5,10\n",
    "\n",
    "for chr in Chr{1..8}\n",
    "do\n",
    "    for SnpGap in 0 5 10\n",
    "    do\n",
    "        echo $chr SnpGap$SnpGap\n",
    "        sbatch -J ${chr}-filter2-removeINDELs-SnpGap${SnpGap} ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_filter2-removeINDELs.sbatch \\\n",
    "            ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_allSNPs+allINDELs.vcf.gz \\\n",
    "            ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_allSNPs.vcf.gz \\\n",
    "            $threads \\\n",
    "            $SnpGap\n",
    "    done\n",
    "done\n",
    "\n",
    "## Run bcftools stats\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/3-stats\n",
    "for inVCF in ~/snap_hap/variants/vcf_bcftools_Am_all/*/Am_all_bcftools_*_allSNPs.vcf.gz\n",
    "do\n",
    "    echo $inVCF\n",
    "    sbatch -J vcf-stats ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_stats.sbatch $inVCF $threads\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aef51d90-593d-4bc5-9abf-f9a4fb4734f8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Filter3: Remove monomoprhic SNPs"
   ]
  },
  {
   "cell_type": "raw",
   "id": "98c9e283-43dc-4364-a52d-b9a68dabc2f1",
   "metadata": {},
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/4-variant-filtering\n",
    "#chr=Chr1\n",
    "#threads=10\n",
    "#SnpGap=10\n",
    "\n",
    "for chr in Chr{1..8}\n",
    "do\n",
    "    for SnpGap in 0 5 10\n",
    "    do\n",
    "        sbatch -J ${chr}-filter3-removeMonoSNPs ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_filter3-removeMonoSNPs.sbatch \\\n",
    "        ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_allSNPs.vcf.gz \\\n",
    "        ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_polySNPs.vcf.gz \\\n",
    "        $threads\n",
    "    done\n",
    "done\n",
    "\n",
    "## Run bcftools stats\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/3-stats\n",
    "for inVCF in ~/snap_hap/variants/vcf_bcftools_Am_all/*/Am_all_bcftools_*_polySNPs.vcf.gz\n",
    "do\n",
    "    echo $inVCF\n",
    "    sbatch -J vcf-stats ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_stats.sbatch $inVCF $threads\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca50eb0e-27a2-4643-8b9c-e3865245e6c3",
   "metadata": {},
   "source": [
    "### Filter4: Remove multiallelic SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5d9d7ad-2e9e-43cb-bdb6-4e9b84fe3839",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/4-variant-filtering\n",
    "#chr=Chr1\n",
    "#threads=10\n",
    "#SnpGap=10\n",
    "\n",
    "for chr in Chr{1..8}\n",
    "do\n",
    "    for SnpGap in 0 5 10\n",
    "    do \n",
    "        echo $chr SnpGap$SnpGap\n",
    "        sbatch -J ${chr}-filter4-removeMultiSNPs ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_filter4-removeMultiSNPs.sbatch \\\n",
    "            ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_polySNPs.vcf.gz \\\n",
    "            ~/snap_hap/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_biSNPs.vcf.gz \\\n",
    "            $threads\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "## Run bcftools stats\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/3-stats\n",
    "threads=10\n",
    "for inVCF in ~/snap_hap/variants/vcf_bcftools_Am_all/*/Am_all_bcftools_*_biSNPs.vcf.gz\n",
    "do\n",
    "    echo $inVCF\n",
    "    sbatch -J vcf-stats ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_stats.sbatch $inVCF $threads\n",
    "done\n",
    "\n",
    "\n",
    "## Run bcftools query to extract params\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/5-extract-params\n",
    "threads=10\n",
    "for inVCF in ~/snap_hap/variants/vcf_bcftools_Am_all/*/Am_all_bcftools_*_SnpGap*_biSNPs.vcf.gz\n",
    "do\n",
    "    echo $inVCF\n",
    "    sbatch -J extract-params-biSNPs ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_extract-params.sbatch $inVCF\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c94caf7-aac2-4820-885b-b1ee31887e3a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<br/>\n",
    "\n",
    "#### Checking filtered variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f27803b2-3acc-4e85-8d80-fcb6c0fe7b34",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1\t955185\tC\tA\t26533.4\t2130\t853\t1246\t0\t42\n",
      "Chr1\t45535107\tC\tA\t28863.3\t2464\t836\t1336\t0\t54\n",
      "Chr1\t47381180\tC\tA\t16317.9\t2576\t314\t1330\t0\t52\n",
      "Chr1\t51252783\tA\tC\t18432.5\t2946\t388\t1364\t0\t53\n",
      "Chr1\t60537705\tT\tC\t14255.6\t2498\t311\t1304\t0\t46\n",
      "Chr1\t71126303\tT\tG\t34834.4\t2830\t909\t1432\t0\t54\n",
      "Chr2\t55473945\tA\tG\t22973.6\t2175\t445\t1324\t0\t57\n",
      "Chr2\t58751341\tA\tG\t8338.71\t1981\t221\t1110\t0\t49\n",
      "Chr3\t1055959\tT\tA\t25268.3\t2470\t813\t1316\t0\t41\n",
      "Chr3\t3488331\tC\tT\t29029.6\t2497\t845\t1350\t0\t48\n",
      "Chr3\t44197469\tG\tC\t18392.9\t2598\t368\t1368\t0\t50\n",
      "Chr4\t7115521\tT\tG\t32824.4\t2682\t948\t1408\t0\t48\n",
      "Chr4\t38397179\tT\tA\t20768.1\t1917\t471\t1214\t0\t54\n",
      "Chr4\t44460308\tA\tG\t26240.1\t3136\t462\t1494\t0\t55\n",
      "Chr4\t50972302\tG\tA\t30206.3\t2642\t831\t1364\t0\t53\n",
      "Chr6\t15125731\tA\tT\t33795.9\t2614\t965\t1440\t0\t53\n",
      "Chr6\t32503688\tA\tG\t25119.7\t2748\t457\t1398\t0\t54\n",
      "Chr6\t52921613\tC\tG\t18310.3\t2466\t369\t1346\t0\t55\n",
      "Chr6\t52922047\tG\tT\t22605\t2515\t458\t1334\t0\t51\n",
      "Chr6\t52966811\tT\tA\t12147\t1920\t313\t1208\t0\t49\n",
      "Chr6\t52999112\tC\tT\t14710.1\t2106\t346\t1200\t0\t50\n",
      "Chr6\t53016551\tC\tG\t18626.9\t2535\t318\t1364\t0\t57\n",
      "Chr6\t53061080\tA\tG\t22587.6\t2113\t482\t1258\t0\t50\n",
      "Chr6\t53083229\tG\tA\t14730.8\t2455\t249\t1352\t0\t56\n",
      "Chr6\t53090212\tT\tC\t27083.1\t2081\t870\t1338\t0\t55\n",
      "Chr6\t53094790\tC\tA\t14456.7\t2600\t239\t1396\t0\t55\n",
      "Chr6\t53104387\tA\tC\t19319\t2586\t433\t1358\t0\t53\n",
      "Chr6\t53195127\tC\tT\t41505.6\t2618\t1089\t1410\t0\t55\n",
      "Chr6\t53944185\tG\tA\t11483.1\t2610\t197\t1400\t0\t52\n",
      "Chr7\t49828363\tT\tA\t18419.7\t3023\t426\t1434\t0\t48\n",
      "Chr8\t27807304\tT\tG\t17404.8\t2532\t381\t1390\t0\t50\n",
      "Chr8\t31864104\tC\tG\t41056.8\t2686\t991\t1414\t0\t56\n",
      "Chr8\t38142268\tC\tG\t27897.2\t2567\t546\t1376\t0\t56\n",
      "Chr8\t44933278\tC\tT\t726.422\t1955\t26\t1242\t0\t33\n",
      "Chr8\t53715058\tC\tT\t25028.9\t2617\t498\t1378\t0\t51\n"
     ]
    }
   ],
   "source": [
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr1/Am_all_bcftools_Chr1_SnpGap5_biSNPs_params.txt | \\\n",
    "    grep -w -e 955185 \\\n",
    "            -e 45535107 \\\n",
    "            -e 47381180 \\\n",
    "            -e 51252783 \\\n",
    "            -e 60537705 \\\n",
    "            -e 71126303\n",
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr2/Am_all_bcftools_Chr2_SnpGap5_biSNPs_params.txt |  grep -w -e 55473945 -e 58751341\n",
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr3/Am_all_bcftools_Chr3_SnpGap5_biSNPs_params.txt | grep -w -e 1055959 -e 3488331 -e 44197469\n",
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr4/Am_all_bcftools_Chr4_SnpGap5_biSNPs_params.txt | grep -w -e 7115521 -e 38397179 -e 44460308 -e 50972302\n",
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr6/Am_all_bcftools_Chr6_SnpGap5_biSNPs_params.txt | grep -w -e 15125731 -e 32503688 -e 52921613 -e 52922047 \\\n",
    "    -e 52966811 -e 52999112 -e 53016551 -e 53061080 -e 53083229 -e 53090212 -e 53094790 -e 53104387 -e 53195127 -e 53944185\n",
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr7/Am_all_bcftools_Chr7_SnpGap5_biSNPs_params.txt | grep -w -e 49828363\n",
    "cat ~/snap_hap/variants/vcf_bcftools_Am_all/Chr8/Am_all_bcftools_Chr8_SnpGap5_biSNPs_params.txt | grep -w -e 27807304 -e 31864104 -e 38142268 -e 44933278 -e 53715058"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "617dfcda-e56c-4b78-94c1-25bc396ee040",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<br>\n",
    "\n",
    "### Filter5: Filter SNPs based on QUAL, DEPTH and MQ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "946ca043-afa4-41d4-8fc9-91c94d5f0e0a",
   "metadata": {},
   "source": [
    "Only SnpGap5 is retained. \n",
    "\n",
    "\n",
    "The average read depth for all samples (n=1074) is 3.6. So, maximum depth allowed = 3.6x1074x2 = 7732.8  \n",
    "Theoretically, the total expected depth should be -  (130x2)+(120x1)+(30x6)+(20x12)+(12x152)+(2x787)+(1x48)+(1x38)+(4x15)+(4x13) = 4396\n",
    "\n",
    "`bcftools filter --threads $threads -e \"INFO/DP<500 | INFO/DP>7732 | QUAL<20 | MQ<30\" -Oz -o $outVCF $inVCF`"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6bda51c7-89ee-41c7-be04-ad29c584fd9f",
   "metadata": {},
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/4-variant-filtering\n",
    "WORKDIR=~/snap_hap\n",
    "threads=10\n",
    "SnpGap=5\n",
    "\n",
    "for chr in Chr{1..8}\n",
    "do\n",
    "    for SnpGap in 5\n",
    "    do \n",
    "        echo $chr SnpGap$SnpGap\n",
    "        sbatch -J ${chr}-filter5-filterSNPs $WORKDIR/_scripts/sbatch/vcf-utils/job-bcftools_filter5-DP_QUAL_MQ.sbatch \\\n",
    "            $WORKDIR/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_biSNPs.vcf.gz \\\n",
    "            $WORKDIR/variants/vcf_bcftools_Am_all/${chr}/Am_all_bcftools_${chr}_SnpGap${SnpGap}_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz \\\n",
    "            $threads\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "## Run bcftools stats\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/bcftools/3-stats\n",
    "threads=10\n",
    "for inVCF in ~/snap_hap/variants/vcf_bcftools_Am_all/*/Am_all_bcftools_*_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "do\n",
    "    echo $inVCF\n",
    "    sbatch -J vcf-stats ~/snap_hap/_scripts/sbatch/vcf-utils/job-bcftools_stats.sbatch $inVCF $threads\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e56ef98-a340-49b7-8a42-867cbaad9284",
   "metadata": {},
   "source": [
    "## No. of SNPs per sample"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e2728958-650f-42fa-b597-47c894699eff",
   "metadata": {},
   "source": [
    "CHROM\tbiSNPs\tfiltered_biSNPs\n",
    "LG1     5962580\t3781002\n",
    "LG2\t    6625191\t4287044\n",
    "LG3\t    5547817\t3400629\n",
    "LG4\t    4444178\t3059258\n",
    "LG5\t    6441785\t4295411\n",
    "LG6\t    4716478\t3361343\n",
    "LG7\t    4568689\t3079991\n",
    "LG8\t    5395028\t3692065"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8818113e-2171-4368-997d-d1a0edd3d2fb",
   "metadata": {},
   "source": [
    "## Extract mpileup genotypes for validation against 30 KASP markers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32d1660d-feca-49d4-ac12-8fd3d3118368",
   "metadata": {},
   "source": [
    "### Chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "61940bda-04c4-4cd5-8647-00114436c868",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr1\t1_1\t955185\t1-Chr1-955185\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.108s\n",
      "user\t0m0.098s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_1-Chr1-955185.noFilter.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 955187\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr1-955185_noFilter\n",
      "\t--from-bp 955183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.087s\n",
      "user\t0m0.078s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_1-Chr1-955185.DP6.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 955187\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr1-955185_DP6\n",
      "\t--from-bp 955183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.108s\n",
      "user\t0m0.101s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_1-Chr1-955185.DP10.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 955187\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr1-955185_DP10\n",
      "\t--from-bp 955183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr1\t1_2\t45535107\t2-Chr1-45535107\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.053s\n",
      "user\t0m0.045s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_2-Chr1-45535107.noFilter.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 45535109\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr1-45535107_noFilter\n",
      "\t--from-bp 45535105\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.051s\n",
      "user\t0m0.040s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_2-Chr1-45535107.DP6.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 45535109\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr1-45535107_DP6\n",
      "\t--from-bp 45535105\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.066s\n",
      "user\t0m0.052s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_2-Chr1-45535107.DP10.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 45535109\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr1-45535107_DP10\n",
      "\t--from-bp 45535105\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr1\t1_3\t47381180\t3-Chr1-47381180\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.116s\n",
      "user\t0m0.101s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_3-Chr1-47381180.noFilter.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 47381182\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr1-47381180_noFilter\n",
      "\t--from-bp 47381178\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.113s\n",
      "user\t0m0.102s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_3-Chr1-47381180.DP6.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 47381182\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr1-47381180_DP6\n",
      "\t--from-bp 47381178\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.105s\n",
      "user\t0m0.098s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_3-Chr1-47381180.DP10.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 47381182\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr1-47381180_DP10\n",
      "\t--from-bp 47381178\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr1\t1_4\t51252783\t4-Chr1-51252783\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.028s\n",
      "user\t0m0.026s\n",
      "sys\t0m0.000s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_4-Chr1-51252783.noFilter.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 51252785\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr1-51252783_noFilter\n",
      "\t--from-bp 51252781\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.027s\n",
      "user\t0m0.018s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_4-Chr1-51252783.DP6.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 51252785\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr1-51252783_DP6\n",
      "\t--from-bp 51252781\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.053s\n",
      "user\t0m0.028s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_4-Chr1-51252783.DP10.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 51252785\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr1-51252783_DP10\n",
      "\t--from-bp 51252781\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr1\t1_5\t60537705\t5-Chr1-60537705\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.190s\n",
      "user\t0m0.169s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_5-Chr1-60537705.noFilter.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 60537707\n",
      "\t--012\n",
      "\t--out ./validation/5-Chr1-60537705_noFilter\n",
      "\t--from-bp 60537703\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.239s\n",
      "user\t0m0.227s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_5-Chr1-60537705.DP6.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 60537707\n",
      "\t--012\n",
      "\t--out ./validation/5-Chr1-60537705_DP6\n",
      "\t--from-bp 60537703\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.239s\n",
      "user\t0m0.225s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_5-Chr1-60537705.DP10.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 60537707\n",
      "\t--012\n",
      "\t--out ./validation/5-Chr1-60537705_DP10\n",
      "\t--from-bp 60537703\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr1\t1_6\t71126303\t6-Chr1-71126303\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.039s\n",
      "user\t0m0.026s\n",
      "sys\t0m0.011s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_6-Chr1-71126303.noFilter.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 71126305\n",
      "\t--012\n",
      "\t--out ./validation/6-Chr1-71126303_noFilter\n",
      "\t--from-bp 71126301\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.049s\n",
      "user\t0m0.038s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_6-Chr1-71126303.DP6.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 71126305\n",
      "\t--012\n",
      "\t--out ./validation/6-Chr1-71126303_DP6\n",
      "\t--from-bp 71126301\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.048s\n",
      "user\t0m0.037s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-1_6-Chr1-71126303.DP10.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 71126305\n",
      "\t--012\n",
      "\t--out ./validation/6-Chr1-71126303_DP10\n",
      "\t--from-bp 71126301\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr1\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in {1..6}\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dade42d-2e0b-421f-8598-525c0a4dbe10",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Chr2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "32bc8ea3-b192-4c59-824a-e2d8299b22f7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr2\t2_1\t55473945\t1-Chr2-55473945\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.088s\n",
      "user\t0m0.074s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-2_1-Chr2-55473945.noFilter.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 55473947\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr2-55473945_noFilter\n",
      "\t--from-bp 55473943\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.090s\n",
      "user\t0m0.080s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-2_1-Chr2-55473945.DP6.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 55473947\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr2-55473945_DP6\n",
      "\t--from-bp 55473943\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.088s\n",
      "user\t0m0.078s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-2_1-Chr2-55473945.DP10.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 55473947\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr2-55473945_DP10\n",
      "\t--from-bp 55473943\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr2\t2_2\t58751341\t2-Chr2-58751341\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.178s\n",
      "user\t0m0.168s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-2_2-Chr2-58751341.noFilter.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 58751343\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr2-58751341_noFilter\n",
      "\t--from-bp 58751339\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.184s\n",
      "user\t0m0.168s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-2_2-Chr2-58751341.DP6.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 58751343\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr2-58751341_DP6\n",
      "\t--from-bp 58751339\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.142s\n",
      "user\t0m0.131s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-2_2-Chr2-58751341.DP10.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 58751343\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr2-58751341_DP10\n",
      "\t--from-bp 58751339\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr2\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in {1..2}\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a66097c7-bb3b-4f8e-9257-865928f0aede",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Chr3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b665da08-624b-4d30-8e8c-500c82df77bf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr3\t3_1\t1055959\t1-Chr3-1055959\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.097s\n",
      "user\t0m0.091s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_1-Chr3-1055959.noFilter.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 1055961\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr3-1055959_noFilter\n",
      "\t--from-bp 1055957\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.088s\n",
      "user\t0m0.082s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_1-Chr3-1055959.DP6.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 1055961\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr3-1055959_DP6\n",
      "\t--from-bp 1055957\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.088s\n",
      "user\t0m0.078s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_1-Chr3-1055959.DP10.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 1055961\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr3-1055959_DP10\n",
      "\t--from-bp 1055957\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr3\t3_2\t3488331\t2-Chr3-3488331\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.186s\n",
      "user\t0m0.175s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_2-Chr3-3488331.noFilter.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 3488333\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr3-3488331_noFilter\n",
      "\t--from-bp 3488329\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.184s\n",
      "user\t0m0.178s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_2-Chr3-3488331.DP6.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 3488333\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr3-3488331_DP6\n",
      "\t--from-bp 3488329\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.229s\n",
      "user\t0m0.210s\n",
      "sys\t0m0.017s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_2-Chr3-3488331.DP10.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 3488333\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr3-3488331_DP10\n",
      "\t--from-bp 3488329\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr3\t3_3\t44197469\t3-Chr3-44197469\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.172s\n",
      "user\t0m0.144s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_3-Chr3-44197469.noFilter.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 44197471\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr3-44197469_noFilter\n",
      "\t--from-bp 44197467\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.163s\n",
      "user\t0m0.137s\n",
      "sys\t0m0.024s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_3-Chr3-44197469.DP6.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 44197471\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr3-44197469_DP6\n",
      "\t--from-bp 44197467\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.157s\n",
      "user\t0m0.142s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-3_3-Chr3-44197469.DP10.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 44197471\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr3-44197469_DP10\n",
      "\t--from-bp 44197467\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr3\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in {1..3}\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4e052aa-90fb-404d-bd6b-85d5996bd73e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Chr4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d185bb65-5202-4ace-8107-b693eb9e8588",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr4\t4_1\t7115521\t1-Chr4-7115521\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.061s\n",
      "user\t0m0.055s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_1-Chr4-7115521.noFilter.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 7115523\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr4-7115521_noFilter\n",
      "\t--from-bp 7115519\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.061s\n",
      "user\t0m0.048s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_1-Chr4-7115521.DP6.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 7115523\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr4-7115521_DP6\n",
      "\t--from-bp 7115519\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.060s\n",
      "user\t0m0.041s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_1-Chr4-7115521.DP10.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 7115523\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr4-7115521_DP10\n",
      "\t--from-bp 7115519\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr4\t4_3\t44460308\t3-Chr4-44460308\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.145s\n",
      "user\t0m0.131s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_3-Chr4-44460308.noFilter.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 44460310\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr4-44460308_noFilter\n",
      "\t--from-bp 44460306\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.112s\n",
      "user\t0m0.109s\n",
      "sys\t0m0.000s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_3-Chr4-44460308.DP6.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 44460310\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr4-44460308_DP6\n",
      "\t--from-bp 44460306\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.145s\n",
      "user\t0m0.138s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_3-Chr4-44460308.DP10.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 44460310\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr4-44460308_DP10\n",
      "\t--from-bp 44460306\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr4\t4_4\t50972302\t4-Chr4-50972302\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.036s\n",
      "user\t0m0.026s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_4-Chr4-50972302.noFilter.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 50972304\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr4-50972302_noFilter\n",
      "\t--from-bp 50972300\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.037s\n",
      "user\t0m0.023s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_4-Chr4-50972302.DP6.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 50972304\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr4-50972302_DP6\n",
      "\t--from-bp 50972300\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.034s\n",
      "user\t0m0.027s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-4_4-Chr4-50972302.DP10.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 50972304\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr4-50972302_DP10\n",
      "\t--from-bp 50972300\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr4\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in {1..3}\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb58953f-4e04-43fe-b456-86677eaf4e48",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Chr6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "86736a91-fbdd-4024-9320-e555a3f3439b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr6\t6_1\t15125731\t1-Chr6-15125731\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.043s\n",
      "user\t0m0.039s\n",
      "sys\t0m0.000s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_1-Chr6-15125731.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 15125733\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr6-15125731_noFilter\n",
      "\t--from-bp 15125729\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.050s\n",
      "user\t0m0.031s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_1-Chr6-15125731.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 15125733\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr6-15125731_DP6\n",
      "\t--from-bp 15125729\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.046s\n",
      "user\t0m0.021s\n",
      "sys\t0m0.022s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_1-Chr6-15125731.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 15125733\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr6-15125731_DP10\n",
      "\t--from-bp 15125729\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_2\t32503688\t2-Chr6-32503688\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.087s\n",
      "user\t0m0.076s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_2-Chr6-32503688.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 32503690\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr6-32503688_noFilter\n",
      "\t--from-bp 32503686\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.086s\n",
      "user\t0m0.072s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_2-Chr6-32503688.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 32503690\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr6-32503688_DP6\n",
      "\t--from-bp 32503686\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.084s\n",
      "user\t0m0.074s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_2-Chr6-32503688.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 32503690\n",
      "\t--012\n",
      "\t--out ./validation/2-Chr6-32503688_DP10\n",
      "\t--from-bp 32503686\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_4\t52922047\t4-Chr6-52922047\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.040s\n",
      "user\t0m0.031s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_4-Chr6-52922047.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52922049\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr6-52922047_noFilter\n",
      "\t--from-bp 52922045\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.038s\n",
      "user\t0m0.024s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_4-Chr6-52922047.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52922049\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr6-52922047_DP6\n",
      "\t--from-bp 52922045\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.040s\n",
      "user\t0m0.025s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_4-Chr6-52922047.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52922049\n",
      "\t--012\n",
      "\t--out ./validation/4-Chr6-52922047_DP10\n",
      "\t--from-bp 52922045\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_5\t52966811\t5-Chr6-52966811\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.125s\n",
      "user\t0m0.119s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_5-Chr6-52966811.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52966813\n",
      "\t--012\n",
      "\t--out ./validation/5-Chr6-52966811_noFilter\n",
      "\t--from-bp 52966809\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.109s\n",
      "user\t0m0.094s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_5-Chr6-52966811.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52966813\n",
      "\t--012\n",
      "\t--out ./validation/5-Chr6-52966811_DP6\n",
      "\t--from-bp 52966809\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.129s\n",
      "user\t0m0.115s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_5-Chr6-52966811.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52966813\n",
      "\t--012\n",
      "\t--out ./validation/5-Chr6-52966811_DP10\n",
      "\t--from-bp 52966809\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_6\t52999112\t6-Chr6-52999112\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.239s\n",
      "user\t0m0.223s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_6-Chr6-52999112.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52999114\n",
      "\t--012\n",
      "\t--out ./validation/6-Chr6-52999112_noFilter\n",
      "\t--from-bp 52999110\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.182s\n",
      "user\t0m0.175s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_6-Chr6-52999112.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52999114\n",
      "\t--012\n",
      "\t--out ./validation/6-Chr6-52999112_DP6\n",
      "\t--from-bp 52999110\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.184s\n",
      "user\t0m0.177s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_6-Chr6-52999112.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52999114\n",
      "\t--012\n",
      "\t--out ./validation/6-Chr6-52999112_DP10\n",
      "\t--from-bp 52999110\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_7\t53016551\t7-Chr6-53016551\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.179s\n",
      "user\t0m0.168s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_7-Chr6-53016551.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53016553\n",
      "\t--012\n",
      "\t--out ./validation/7-Chr6-53016551_noFilter\n",
      "\t--from-bp 53016549\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.227s\n",
      "user\t0m0.225s\n",
      "sys\t0m0.001s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_7-Chr6-53016551.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53016553\n",
      "\t--012\n",
      "\t--out ./validation/7-Chr6-53016551_DP6\n",
      "\t--from-bp 53016549\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.228s\n",
      "user\t0m0.213s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_7-Chr6-53016551.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53016553\n",
      "\t--012\n",
      "\t--out ./validation/7-Chr6-53016551_DP10\n",
      "\t--from-bp 53016549\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_8\t53061080\t8-Chr6-53061080\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.096s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_8-Chr6-53061080.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53061082\n",
      "\t--012\n",
      "\t--out ./validation/8-Chr6-53061080_noFilter\n",
      "\t--from-bp 53061078\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.114s\n",
      "user\t0m0.096s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_8-Chr6-53061080.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53061082\n",
      "\t--012\n",
      "\t--out ./validation/8-Chr6-53061080_DP6\n",
      "\t--from-bp 53061078\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.110s\n",
      "user\t0m0.100s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_8-Chr6-53061080.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53061082\n",
      "\t--012\n",
      "\t--out ./validation/8-Chr6-53061080_DP10\n",
      "\t--from-bp 53061078\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_9\t53083229\t9-Chr6-53083229\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.184s\n",
      "user\t0m0.181s\n",
      "sys\t0m0.000s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_9-Chr6-53083229.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53083231\n",
      "\t--012\n",
      "\t--out ./validation/9-Chr6-53083229_noFilter\n",
      "\t--from-bp 53083227\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.183s\n",
      "user\t0m0.169s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_9-Chr6-53083229.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53083231\n",
      "\t--012\n",
      "\t--out ./validation/9-Chr6-53083229_DP6\n",
      "\t--from-bp 53083227\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.197s\n",
      "user\t0m0.167s\n",
      "sys\t0m0.016s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_9-Chr6-53083229.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53083231\n",
      "\t--012\n",
      "\t--out ./validation/9-Chr6-53083229_DP10\n",
      "\t--from-bp 53083227\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_10\t53090212\t10-Chr6-53090212\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.092s\n",
      "user\t0m0.082s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_10-Chr6-53090212.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53090214\n",
      "\t--012\n",
      "\t--out ./validation/10-Chr6-53090212_noFilter\n",
      "\t--from-bp 53090210\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.090s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_10-Chr6-53090212.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53090214\n",
      "\t--012\n",
      "\t--out ./validation/10-Chr6-53090212_DP6\n",
      "\t--from-bp 53090210\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.093s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_10-Chr6-53090212.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53090214\n",
      "\t--012\n",
      "\t--out ./validation/10-Chr6-53090212_DP10\n",
      "\t--from-bp 53090210\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_11\t53094790\t11-Chr6-53094790\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.150s\n",
      "user\t0m0.144s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_11-Chr6-53094790.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53094792\n",
      "\t--012\n",
      "\t--out ./validation/11-Chr6-53094790_noFilter\n",
      "\t--from-bp 53094788\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.154s\n",
      "user\t0m0.151s\n",
      "sys\t0m0.001s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_11-Chr6-53094790.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53094792\n",
      "\t--012\n",
      "\t--out ./validation/11-Chr6-53094790_DP6\n",
      "\t--from-bp 53094788\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.152s\n",
      "user\t0m0.145s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_11-Chr6-53094790.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53094792\n",
      "\t--012\n",
      "\t--out ./validation/11-Chr6-53094790_DP10\n",
      "\t--from-bp 53094788\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_12\t53104387\t12-Chr6-53104387\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.064s\n",
      "user\t0m0.058s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_12-Chr6-53104387.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53104389\n",
      "\t--012\n",
      "\t--out ./validation/12-Chr6-53104387_noFilter\n",
      "\t--from-bp 53104385\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.078s\n",
      "user\t0m0.068s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_12-Chr6-53104387.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53104389\n",
      "\t--012\n",
      "\t--out ./validation/12-Chr6-53104387_DP6\n",
      "\t--from-bp 53104385\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.080s\n",
      "user\t0m0.074s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_12-Chr6-53104387.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53104389\n",
      "\t--012\n",
      "\t--out ./validation/12-Chr6-53104387_DP10\n",
      "\t--from-bp 53104385\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_13\t53195127\t13-Chr6-53195127\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.174s\n",
      "user\t0m0.160s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_13-Chr6-53195127.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53195129\n",
      "\t--012\n",
      "\t--out ./validation/13-Chr6-53195127_noFilter\n",
      "\t--from-bp 53195125\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.171s\n",
      "user\t0m0.161s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_13-Chr6-53195127.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53195129\n",
      "\t--012\n",
      "\t--out ./validation/13-Chr6-53195127_DP6\n",
      "\t--from-bp 53195125\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.174s\n",
      "user\t0m0.152s\n",
      "sys\t0m0.020s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_13-Chr6-53195127.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53195129\n",
      "\t--012\n",
      "\t--out ./validation/13-Chr6-53195127_DP10\n",
      "\t--from-bp 53195125\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr6\t6_14\t53944185\t14-Chr6-53944185\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.144s\n",
      "user\t0m0.133s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_14-Chr6-53944185.noFilter.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53944187\n",
      "\t--012\n",
      "\t--out ./validation/14-Chr6-53944185_noFilter\n",
      "\t--from-bp 53944183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.155s\n",
      "user\t0m0.129s\n",
      "sys\t0m0.012s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_14-Chr6-53944185.DP6.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53944187\n",
      "\t--012\n",
      "\t--out ./validation/14-Chr6-53944185_DP6\n",
      "\t--from-bp 53944183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.145s\n",
      "user\t0m0.130s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-6_14-Chr6-53944185.DP10.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53944187\n",
      "\t--012\n",
      "\t--out ./validation/14-Chr6-53944185_DP10\n",
      "\t--from-bp 53944183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr6\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in {1..13}\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35f30da4-32e1-4bf2-9a3a-131c16d11dad",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Chr7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2eb94acb-f1ef-49c6-9141-9ee237ad69ab",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr7\t7_1\t49828363\t1-Chr7-49828363\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.107s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.013s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-7_1-Chr7-49828363.noFilter.vcf.gz\n",
      "\t--chr Chr7\n",
      "\t--to-bp 49828365\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr7-49828363_noFilter\n",
      "\t--from-bp 49828361\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.085s\n",
      "user\t0m0.075s\n",
      "sys\t0m0.009s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-7_1-Chr7-49828363.DP6.vcf.gz\n",
      "\t--chr Chr7\n",
      "\t--to-bp 49828365\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr7-49828363_DP6\n",
      "\t--from-bp 49828361\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.108s\n",
      "user\t0m0.098s\n",
      "sys\t0m0.008s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-7_1-Chr7-49828363.DP10.vcf.gz\n",
      "\t--chr Chr7\n",
      "\t--to-bp 49828365\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr7-49828363_DP10\n",
      "\t--from-bp 49828361\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr7\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in 1\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecf422c-1d26-452e-9b6f-3bdcd911c175",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Chr8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8513fd8a-f974-4994-be1b-ffe0411eed28",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr8\t8_1\t27807304\t1-Chr8-27807304\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.048s\n",
      "user\t0m0.043s\n",
      "sys\t0m0.004s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-8_1-Chr8-27807304.noFilter.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 27807306\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr8-27807304_noFilter\n",
      "\t--from-bp 27807302\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.052s\n",
      "user\t0m0.046s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-8_1-Chr8-27807304.DP6.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 27807306\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr8-27807304_DP6\n",
      "\t--from-bp 27807302\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.064s\n",
      "user\t0m0.041s\n",
      "sys\t0m0.010s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-8_1-Chr8-27807304.DP10.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 27807306\n",
      "\t--012\n",
      "\t--out ./validation/1-Chr8-27807304_DP10\n",
      "\t--from-bp 27807302\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n",
      "\n",
      "Chr8\t8_3\t38142268\t3-Chr8-38142268\n",
      "\n",
      "\n",
      "No Filter\n",
      "\n",
      "real\t0m0.024s\n",
      "user\t0m0.011s\n",
      "sys\t0m0.011s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-8_3-Chr8-38142268.noFilter.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 38142270\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr8-38142268_noFilter\n",
      "\t--from-bp 38142266\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<6\n",
      "\n",
      "real\t0m0.027s\n",
      "user\t0m0.020s\n",
      "sys\t0m0.005s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-8_3-Chr8-38142268.DP6.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 38142270\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr8-38142268_DP6\n",
      "\t--from-bp 38142266\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Softmask FORMAT/DP<10\n",
      "\n",
      "real\t0m0.025s\n",
      "user\t0m0.009s\n",
      "sys\t0m0.014s\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf ./validation/mpileup-8_3-Chr8-38142268.DP10.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 38142270\n",
      "\t--012\n",
      "\t--out ./validation/3-Chr8-38142268_DP10\n",
      "\t--from-bp 38142266\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create combined GT file\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/vcf_bcftools_Am_all\n",
    "KASPcoords=./KASPcoords.pos\n",
    "\n",
    "chrom=Chr8\n",
    "vcf=./${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "outDIR=./validation\n",
    "\n",
    "for snpID in {1..2}\n",
    "do\n",
    "    markerID=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f3)\n",
    "    pos=$(grep -w $chrom $KASPcoords | sed -n \"${snpID}p\" | cut -f4)\n",
    "    outVCF=mpileup-${markerID}-${chrom}-${pos}\n",
    "    outGT=${markerID#*_}-${chrom}-${pos}\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$markerID\"\\t\"$pos\"\\t\"$outGT\"\\n\"\n",
    "    \n",
    "    ## No Filter\n",
    "    echo -e '\\nNo Filter'\n",
    "    time bcftools view -r ${chrom}:${pos} -Oz -o $outDIR/$outVCF.noFilter.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.noFilter.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.noFilter.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_noFilter\n",
    "\n",
    "    ## Softmask FORMAT/DP<6 \n",
    "    echo -e '\\nSoftmask FORMAT/DP<6'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<6\" -Oz -o $outDIR/$outVCF.DP6.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP6.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP6.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP6\n",
    "\n",
    "    ## Softmask FORMAT/DP<10\n",
    "    echo -e '\\nSoftmask FORMAT/DP<10'\n",
    "    time bcftools filter -r ${chrom}:${pos} -S . -e \"FMT/DP<10\" -Oz -o $outDIR/$outVCF.DP10.vcf.gz $vcf\n",
    "    bcftools tabix -f $outDIR/$outVCF.DP10.vcf.gz\n",
    "    vcftools --gzvcf $outDIR/$outVCF.DP10.vcf.gz --chr $chrom --from-bp $((pos-2)) --to-bp $((pos+2)) --012 --out $outDIR/${outGT}_DP10\n",
    "\n",
    "    ## Change longFormat sample names to shortFormat\n",
    "    echo -e '\\nChange longFormat sample names to shortFormat'\n",
    "    cat $outDIR/${markerID#*_}-${chrom}-${pos}_DP10.012.indv | \\\n",
    "        sed 's/\\(10x_\\|10x2_\\|10xNEW_\\|Pla_\\|Ave_\\|GH_\\|60x_\\|60x2_\\|2x_\\)//g; s/\\(TRIO-\\|n96-\\)//g' > $outDIR/${outGT}.shortFormat.indv\n",
    "\n",
    "    ## Create file with stitch genotypes\n",
    "    echo -e '\\nCreate combined GT file'\n",
    "    paste $outDIR/${outGT}.shortFormat.indv \\\n",
    "        <(cut -f2 $outDIR/${outGT}_noFilter.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP6.012) \\\n",
    "        <(cut -f2 $outDIR/${outGT}_DP10.012) > $outDIR/mpileup_$outGT.gt\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
