{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "df3380d2-24c6-4847-98f9-ca2f54bfe882",
   "metadata": {},
   "source": [
    "# Postprocessing STITCH output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48f19054-1281-4b19-b70c-af9d326e3783",
   "metadata": {},
   "source": [
    "- Step 1 - Extract all genotypes from high-confidence STITCH output VCF based on posterior genotype prob.  \n",
    "- Step 2 - Merge all chrom_segment VCFs into one VCF per chromosome  \n",
    "- Step 3 - Extract VCF stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "25a1c293-b819-4d31-a121-8830ee862135",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Currently Loaded Modules:\n",
      "  1) git/2.35.1         7) R/4.3.1-b                  13) GSL/2.7\n",
      "  2) java/17            8) rstudio/2023.06.1-524      14) htslib/1.18\n",
      "  3) texlive/20211108   9) nodejs/18.16.1        (T)  15) perl/5.36.0\n",
      "  4) hdf5/1.14.2       10) julia/1.10.1               16) bcftools/1.18\n",
      "  5) gdal/3.4.2        11) proj/8.1.1                 17) vcftools/0.1.16\n",
      "  6) glpk/5.0          12) JupyterPython/2.0     (T)  18) parallel/20220222\n",
      "\n",
      "  Where:\n",
      "   T:  Testing\n",
      "\n",
      " \n",
      "\n"
     ]
    }
   ],
   "source": [
    "module load bcftools/1.18 vcftools/0.1.16 parallel\n",
    "module list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bbb813a-cff2-4fbf-83d2-5c9ddb8a5267",
   "metadata": {},
   "source": [
    "## Steps for postprocessing STITCH VCF output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1224c067-98eb-41fe-a417-ebbf33704670",
   "metadata": {},
   "source": [
    "*NB: Change the baseDIR for stitch Run 2.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8517732d-1bcd-4905-85e4-62e2be9dd408",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Initiate variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "64f1fe7c-2d58-4971-8cb6-726facf56386",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "chrom: Chr6\n",
      "baseDIR: /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr6\n",
      "outVCF: /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr6/Am_all_stitch_Chr6_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Input variable\n",
    "chrom=Chr6\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "\n",
    "echo -e \"\\n\"\n",
    "echo chrom: $chrom\n",
    "echo baseDIR: $baseDIR \n",
    "echo -e '\\n'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "495ea59e-6299-4bd4-8d60-c622a9e863d3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extract all genotypes from high-confidence stitch output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa6839dc-2ef4-4c1b-9ec3-e22bca90ed4b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "echo -e '\\nExtract all genotypes from STITCH'\n",
    "realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz > ${baseDIR}/stitchVcfs.highConf.list\n",
    "\n",
    "## Execute in parallel mode\n",
    "#time parallel -j10 \"bash ~/snap_hap/_scripts/bash/stitch/add_PG_PL.sh {}\" ::: ${baseDIR}/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz\n",
    "\n",
    "## Execute one-by-one (parallel array job in cluster)\n",
    "SLURM_ARRAY_TASK_ID=1\n",
    "inVCF=$(sed -n \"${SLURM_ARRAY_TASK_ID}p\" $stitchVcfList)\n",
    "echo $inVCF\n",
    "time bash ~/snap_hap/_scripts/bash/stitch/add_PG_PL.sh $inVCF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74831395-dc8f-4eac-85ef-073c90f960ea",
   "metadata": {},
   "source": [
    "### Merge vcfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6190101-aa75-4e79-b9ad-1624b1b4882e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make stitch-highConf-VCF list\n",
    "echo -e '\\nMerge highConf stitch VCFs'\n",
    "\n",
    "realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz > ${baseDIR}/stitchVcfs.highConf.list\n",
    "stitchVcfList_highConf=${baseDIR}/stitchVcfs.highConf.list\n",
    "outVCF_highConf=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "\n",
    "time bcftools concat -a -Oz -o ${outVCF_highConf} -f ${stitchVcfList_highConf}\n",
    "time bcftools stats --threads 10 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 ${outVCF_highConf} > ${outVCF_highConf/.vcf.gz/.stats}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "382a82b7-c93c-4fac-93db-bcd32486e756",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make stitch-ALL-VCF list\n",
    "echo -e '\\nMerge ALL stitch VCFs'\n",
    "\n",
    "realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*.PL.vcf.gz > ${baseDIR}/stitchVcfs.ALL.list\n",
    "stitchVcfList_ALL=${baseDIR}/stitchVcfs.ALL.list\n",
    "outVCF_ALL=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.vcf.gz\n",
    "\n",
    "time bcftools concat -a -Oz -o ${outVCF_ALL} -f ${stitchVcfList_ALL}\n",
    "time bcftools stats --threads 10 -s- --af-bins <(seq 0 0.1 1) --depth 0,25000,25 ${outVCF_ALL} > ${outVCF_ALL/.vcf.gz/.stats}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7005d41b-afe3-4cf9-a550-9c284bf59bc2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Add AC,AN annotations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "627cdff7-8700-4859-b68d-f177c7e6aca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#inputVCF = ...\n",
    "bash ~/snap_hap/_scripts/bash/fill-AC-AN_vcfgzFormat.sh $inputVCF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f753a6dc-c6bc-485d-9306-8287b421dd21",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Create VCF statitics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "597aa47b-c0c3-4d17-9a62-92de1eb13890",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate missing genotypes per site and individual\n",
    "echo -e '\\nCalculating fraction of missing data'\n",
    "time vcftools --gzvcf $outVCF --missing-indv --out ${chrom}_missing-indv\n",
    "time vcftools --gzvcf $outVCF --missing-site --out ${chrom}_missing-site\n",
    "\n",
    "## Calculate heterozygosity per individual\n",
    "echo -e '\\nCalculating heterozygosity'\n",
    "time vcftools --gzvcf $outVCF --het --out ${chrom}_het-indv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f718bca-9c0d-4c6b-ad01-511ad9affb07",
   "metadata": {},
   "source": [
    "## Cluster implementation: postprocessing STITCH output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da8ee083-f1ce-4386-b3a3-8241cab14677",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Make list of high-conf STITCH vcfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ad58c2e-1a31-438f-9efc-74e4ecd59029",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr1\t72 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr1/stitchVcfs.highConf.list\n",
      "\n",
      "Chr2\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr2\t78 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr2/stitchVcfs.highConf.list\n",
      "\n",
      "Chr3\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr3\t66 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr3/stitchVcfs.highConf.list\n",
      "\n",
      "Chr4\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr4\t55 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr4/stitchVcfs.highConf.list\n",
      "\n",
      "Chr5\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr5\t71 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr5/stitchVcfs.highConf.list\n",
      "\n",
      "Chr6\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr6\t56 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr6/stitchVcfs.highConf.list\n",
      "\n",
      "Chr7\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr7\t56 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr7/stitchVcfs.highConf.list\n",
      "\n",
      "Chr8\t/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr8\t58 files\n",
      "/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr8/stitchVcfs.highConf.list\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "    echo -e $chrom'\\t'$baseDIR'\\t'$(realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz | wc -l) files\n",
    "    realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz > ${baseDIR}/stitchVcfs.highConf.list\n",
    "    echo -e ${baseDIR}/stitchVcfs.highConf.list'\\n'\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb49830a-4b71-459b-b806-805c89dad93a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Extract all genotypes from high-confidence stitch output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ac4b494-35a6-4cc0-9355-d6877209fd54",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/postSTITCH/extractGT\n",
    "\n",
    "# Chr1\n",
    "chrom=Chr1\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-72 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr2\n",
    "chrom=Chr2\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-78 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr3\n",
    "chrom=Chr3\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-66 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr4\n",
    "chrom=Chr4\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-55 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr5\n",
    "chrom=Chr5\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-71 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr6\n",
    "chrom=Chr6\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-56 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr7\n",
    "chrom=Chr7\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-56 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list\n",
    "# Chr8\n",
    "chrom=Chr8\n",
    "baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "sbatch --array=1-58 -J postSTITCH-extractGT_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch $chrom ${baseDIR}/stitchVcfs.highConf.list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73751bec-3704-4a26-b637-9fcf68869fb2",
   "metadata": {},
   "source": [
    "```\n",
    "#SBATCH --partition=defaultp\n",
    "#SBATCH --time=00:30:00\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem-per-cpu=4G\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f85e5b84-62f4-4dd3-98c5-4ca211e848fb",
   "metadata": {},
   "source": [
    "It takes slightly longer time if run with `time srun ...` instead of `srun time ...` with the SLURM script. Generally, takes around ~10-15 mins."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a92ac92-ba85-4aa0-9c38-61e86d664a22",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Concat VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e43fbb15-8adb-4f0e-8588-fecc60253c16",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/postSTITCH/concatVcfs\n",
    "\n",
    "for chrom in Chr{1..8}\n",
    "do \n",
    "    baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "    echo -e $chrom'\\t'$baseDIR\n",
    "\n",
    "    #highConf#\n",
    "    realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz > ${baseDIR}/stitchVcfs.highConf.list\n",
    "    stitchVcfList_highConf=${baseDIR}/stitchVcfs.highConf.list\n",
    "    outVCF_highConf=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "    sbatch -J postSTITCH-concatVcfs-highConf_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-concatVcfs.sbatch.sh ${chrom} ${stitchVcfList_highConf} ${outVCF_highConf}\n",
    "\n",
    "    #ALL#\n",
    "    realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*.PL.vcf.gz > ${baseDIR}/stitchVcfs.ALL.list\n",
    "    stitchVcfList_ALL=${baseDIR}/stitchVcfs.ALL.list\n",
    "    outVCF_ALL=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.vcf.gz\n",
    "    sbatch -J postSTITCH-concatVcfs-ALL_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-concatVcfs.sbatch.sh ${chrom} ${stitchVcfList_ALL} ${outVCF_ALL}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73f43d13-e454-4315-8601-9343171b1a78",
   "metadata": {},
   "source": [
    "```\n",
    "#SBATCH --partition=defaultp\n",
    "#SBATCH --time=3:00:00\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --cpus-per-task=5\n",
    "#SBATCH --mem-per-cpu=4G\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46fd39d9-299b-4c35-85c4-ebfa22597c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "User    Account      State        JobID    JobName  AllocCPUS AllocNodes    CPUTime     ReqMem     MaxRSS    Elapsed  AllocTRES \n",
    "--------- ---------- ---------- ------------ ---------- ---------- ---------- ---------- ---------- ---------- ---------- ---------- \n",
    "     apal   bartogrp     FAILED 9568345      postSTITC+          1          1   00:50:34       16Gc              00:50:34 billing=1+ \n",
    "            bartogrp  COMPLETED 9568345.0          time          1          1   00:50:25       16Gc     30228K   00:50:25 cpu=1,mem+"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b907c4-fa44-4d98-89e0-a259f1503904",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Add AC, AN annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c29f4863-afcf-4727-b5a1-4394c5f12a3e",
   "metadata": {},
   "source": [
    "**NB:** Don't run AC,AN fill on cluster due to incompatibilities of bcftools libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39f873ea-a63d-48d2-9e01-69741b5608f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/postSTITCH/annotateVcfs\n",
    "\n",
    "for chrom in Chr{1..8}\n",
    "do \n",
    "    baseDIR=~/snap_hap/variants/stitch/${chrom}\n",
    "    echo -e $chrom'\\t'$baseDIR\n",
    "\n",
    "    #highConf#\n",
    "    inputVCF_highConf=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "    sbatch -J ${chrom}_highConf_tagVCFs ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-fill_AC_AN.sbatch ${inputVCF_highConf}\n",
    "\n",
    "    #ALL#\n",
    "    inputVCF_ALL=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.vcf.gz\n",
    "    sbatch -J ${chrom}_highConf_tagVCFs ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-fill_AC_AN.sbatch ${inputVCF_ALL}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5328d82-e14f-423b-a743-d0b47ca54706",
   "metadata": {},
   "source": [
    "## stitch VCF statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a59e4723-d913-4d35-a1a6-5909687e44cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No. of stitched records\n",
      "Chr1\t3781002\n",
      "Chr2\t4287044\n",
      "Chr3\t3400629\n",
      "Chr4\t3059258\n",
      "Chr5\t4295411\n",
      "Chr6\t3361343\n",
      "Chr7\t3079991\n",
      "Chr8\t3692065\n"
     ]
    }
   ],
   "source": [
    "basedir=\"/nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch\"\n",
    "\n",
    "echo -e 'No. of stitched records'\n",
    "for chrom in Chr{1..8}; do records=`grep \"number of records:\" $basedir/$chrom/*PL.stats | cut -f4`; echo -e $chrom\"\\t\"$records; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb856d18-802d-41cf-8165-5dc4edec029f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract imputed genotypes for validation against 30 KASP markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "468cbeec-7e56-4fee-ac15-89b2bb4a76fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ad01ef2f-7ebe-4476-9ba8-a6ae887f2bcd",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snpID\tCHROM\tPOS\n",
      "1\tChr1\t955185\n",
      "2\tChr1\t45535107\n",
      "3\tChr1\t47381180\n",
      "4\tChr1\t51252783\n",
      "5\tChr1\t60537705\n",
      "6\tChr1\t71126303\n",
      "1\tChr2\t55473945\n",
      "2\tChr2\t58751341\n",
      "1\tChr3\t1055959\n",
      "2\tChr3\t3488331\n",
      "3\tChr3\t44197469\n",
      "1\tChr4\t7115521\n",
      "2\tChr4\t38397179\n",
      "3\tChr4\t44460308\n",
      "4\tChr4\t50972302\n",
      "1\tChr6\t15125731\n",
      "2\tChr6\t32503688\n",
      "3\tChr6\t52921613\n",
      "4\tChr6\t52922047\n",
      "5\tChr6\t52966811\n",
      "6\tChr6\t52999112\n",
      "7\tChr6\t53016551\n",
      "8\tChr6\t53061080\n",
      "9\tChr6\t53083229\n",
      "10\tChr6\t53090212\n",
      "11\tChr6\t53094790\n",
      "12\tChr6\t53104387\n",
      "13\tChr6\t53195127\n",
      "14\tChr6\t53944185\n",
      "1\tChr7\t49828363\n",
      "1\tChr8\t27807304\n",
      "2\tChr8\t31864104\n",
      "3\tChr8\t38142268\n",
      "4\tChr8\t44933278\n",
      "5\tChr8\t53715058\n"
     ]
    }
   ],
   "source": [
    "## Read KASP coords positions\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "cat $KASPcoords | cut -f1,4,8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e70e64b5-77d5-4c4c-b51e-893b45478500",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr1\t1\t1_1\t955185\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.139s\n",
      "user\t0m0.044s\n",
      "sys\t0m0.041s\n",
      "\n",
      "real\t0m0.099s\n",
      "user\t0m0.058s\n",
      "sys\t0m0.036s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr1_955185.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 955187\n",
      "\t--012\n",
      "\t--out 1-Chr1-955185_highConf\n",
      "\t--from-bp 955183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.172s\n",
      "user\t0m0.052s\n",
      "sys\t0m0.121s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr1_955185.PL.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 955187\n",
      "\t--012\n",
      "\t--out 1-Chr1-955185_ALL\n",
      "\t--from-bp 955183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m3.703s\n",
      "user\t0m1.003s\n",
      "sys\t0m1.158s\n",
      "\n",
      "Chr1\t2\t1_2\t45535107\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.057s\n",
      "user\t0m0.030s\n",
      "sys\t0m0.023s\n",
      "\n",
      "real\t0m0.093s\n",
      "user\t0m0.049s\n",
      "sys\t0m0.017s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr1_45535107.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 45535109\n",
      "\t--012\n",
      "\t--out 2-Chr1-45535107_highConf\n",
      "\t--from-bp 45535105\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.159s\n",
      "user\t0m0.064s\n",
      "sys\t0m0.101s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr1_45535107.PL.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 45535109\n",
      "\t--012\n",
      "\t--out 2-Chr1-45535107_ALL\n",
      "\t--from-bp 45535105\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m3.231s\n",
      "user\t0m0.890s\n",
      "sys\t0m0.983s\n",
      "\n",
      "Chr1\t3\t1_3\t47381180\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.093s\n",
      "user\t0m0.058s\n",
      "sys\t0m0.030s\n",
      "\n",
      "real\t0m0.445s\n",
      "user\t0m0.082s\n",
      "sys\t0m0.037s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr1_47381180.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 47381182\n",
      "\t--012\n",
      "\t--out 3-Chr1-47381180_highConf\n",
      "\t--from-bp 47381178\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.156s\n",
      "user\t0m0.053s\n",
      "sys\t0m0.107s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr1_47381180.PL.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 47381182\n",
      "\t--012\n",
      "\t--out 3-Chr1-47381180_ALL\n",
      "\t--from-bp 47381178\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.471s\n",
      "user\t0m0.913s\n",
      "sys\t0m0.854s\n",
      "\n",
      "Chr1\t4\t1_4\t51252783\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.041s\n",
      "user\t0m0.017s\n",
      "sys\t0m0.021s\n",
      "\n",
      "real\t0m0.282s\n",
      "user\t0m0.015s\n",
      "sys\t0m0.030s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr1_51252783.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 51252785\n",
      "\t--012\n",
      "\t--out 4-Chr1-51252783_highConf\n",
      "\t--from-bp 51252781\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.160s\n",
      "user\t0m0.053s\n",
      "sys\t0m0.112s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr1_51252783.PL.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 51252785\n",
      "\t--012\n",
      "\t--out 4-Chr1-51252783_ALL\n",
      "\t--from-bp 51252781\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.548s\n",
      "user\t0m0.838s\n",
      "sys\t0m0.971s\n",
      "\n",
      "Chr1\t5\t1_5\t60537705\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.129s\n",
      "user\t0m0.085s\n",
      "sys\t0m0.039s\n",
      "\n",
      "real\t0m0.487s\n",
      "user\t0m0.156s\n",
      "sys\t0m0.017s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 5_Chr1_60537705.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 60537707\n",
      "\t--012\n",
      "\t--out 5-Chr1-60537705_highConf\n",
      "\t--from-bp 60537703\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.163s\n",
      "user\t0m0.060s\n",
      "sys\t0m0.111s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 5_Chr1_60537705.PL.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 60537707\n",
      "\t--012\n",
      "\t--out 5-Chr1-60537705_ALL\n",
      "\t--from-bp 60537703\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.530s\n",
      "user\t0m0.864s\n",
      "sys\t0m0.979s\n",
      "\n",
      "Chr1\t6\t1_6\t71126303\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.041s\n",
      "user\t0m0.022s\n",
      "sys\t0m0.017s\n",
      "\n",
      "real\t0m0.127s\n",
      "user\t0m0.039s\n",
      "sys\t0m0.016s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 6_Chr1_71126303.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 71126305\n",
      "\t--012\n",
      "\t--out 6-Chr1-71126303_highConf\n",
      "\t--from-bp 71126301\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.157s\n",
      "user\t0m0.066s\n",
      "sys\t0m0.101s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 6_Chr1_71126303.PL.vcf.gz\n",
      "\t--chr Chr1\n",
      "\t--to-bp 71126305\n",
      "\t--012\n",
      "\t--out 6-Chr1-71126303_ALL\n",
      "\t--from-bp 71126301\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.245s\n",
      "user\t0m0.906s\n",
      "sys\t0m0.855s\n"
     ]
    }
   ],
   "source": [
    "## Chr1\n",
    "chrom=Chr1\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in {1..6}\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8ce5283a-b947-4740-9202-fca7efc749f0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr2\t1\t2_1\t55473945\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.060s\n",
      "user\t0m0.033s\n",
      "sys\t0m0.022s\n",
      "\n",
      "real\t0m0.093s\n",
      "user\t0m0.064s\n",
      "sys\t0m0.013s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr2_55473945.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 55473947\n",
      "\t--012\n",
      "\t--out 1-Chr2-55473945_highConf\n",
      "\t--from-bp 55473943\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.188s\n",
      "user\t0m0.053s\n",
      "sys\t0m0.110s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr2_55473945.PL.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 55473947\n",
      "\t--012\n",
      "\t--out 1-Chr2-55473945_ALL\n",
      "\t--from-bp 55473943\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.241s\n",
      "user\t0m0.832s\n",
      "sys\t0m0.931s\n",
      "\n",
      "Chr2\t2\t2_2\t58751341\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.102s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.016s\n",
      "\n",
      "real\t0m0.172s\n",
      "user\t0m0.129s\n",
      "sys\t0m0.025s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr2_58751341.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 58751343\n",
      "\t--012\n",
      "\t--out 2-Chr2-58751341_highConf\n",
      "\t--from-bp 58751339\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.163s\n",
      "user\t0m0.065s\n",
      "sys\t0m0.106s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr2_58751341.PL.vcf.gz\n",
      "\t--chr Chr2\n",
      "\t--to-bp 58751343\n",
      "\t--012\n",
      "\t--out 2-Chr2-58751341_ALL\n",
      "\t--from-bp 58751339\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.523s\n",
      "user\t0m0.858s\n",
      "sys\t0m0.907s\n"
     ]
    }
   ],
   "source": [
    "## Chr2\n",
    "chrom=Chr2\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in {1..2}\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "af083e24-fc9a-475d-871a-164550682267",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr3\t1\t3_1\t1055959\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.194s\n",
      "user\t0m0.058s\n",
      "sys\t0m0.018s\n",
      "\n",
      "real\t0m0.370s\n",
      "user\t0m0.071s\n",
      "sys\t0m0.034s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr3_1055959.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 1055961\n",
      "\t--012\n",
      "\t--out 1-Chr3-1055959_highConf\n",
      "\t--from-bp 1055957\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.177s\n",
      "user\t0m0.054s\n",
      "sys\t0m0.123s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr3_1055959.PL.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 1055961\n",
      "\t--012\n",
      "\t--out 1-Chr3-1055959_ALL\n",
      "\t--from-bp 1055957\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.273s\n",
      "user\t0m0.860s\n",
      "sys\t0m0.930s\n",
      "\n",
      "Chr3\t2\t3_2\t3488331\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.139s\n",
      "user\t0m0.092s\n",
      "sys\t0m0.025s\n",
      "\n",
      "real\t0m0.313s\n",
      "user\t0m0.127s\n",
      "sys\t0m0.020s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr3_3488331.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 3488333\n",
      "\t--012\n",
      "\t--out 2-Chr3-3488331_highConf\n",
      "\t--from-bp 3488329\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.163s\n",
      "user\t0m0.048s\n",
      "sys\t0m0.116s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr3_3488331.PL.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 3488333\n",
      "\t--012\n",
      "\t--out 2-Chr3-3488331_ALL\n",
      "\t--from-bp 3488329\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.639s\n",
      "user\t0m0.894s\n",
      "sys\t0m0.935s\n",
      "\n",
      "Chr3\t3\t3_3\t44197469\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.127s\n",
      "user\t0m0.077s\n",
      "sys\t0m0.033s\n",
      "\n",
      "real\t0m0.466s\n",
      "user\t0m0.133s\n",
      "sys\t0m0.020s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr3_44197469.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 44197471\n",
      "\t--012\n",
      "\t--out 3-Chr3-44197469_highConf\n",
      "\t--from-bp 44197467\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.169s\n",
      "user\t0m0.053s\n",
      "sys\t0m0.118s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr3_44197469.PL.vcf.gz\n",
      "\t--chr Chr3\n",
      "\t--to-bp 44197471\n",
      "\t--012\n",
      "\t--out 3-Chr3-44197469_ALL\n",
      "\t--from-bp 44197467\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m3.013s\n",
      "user\t0m0.933s\n",
      "sys\t0m0.947s\n"
     ]
    }
   ],
   "source": [
    "## Chr2\n",
    "chrom=Chr3\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in {1..3}\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f244ed1f-998b-4f69-8d25-7a79c5319946",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr4\t1\t4_1\t7115521\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.050s\n",
      "user\t0m0.031s\n",
      "sys\t0m0.012s\n",
      "\n",
      "real\t0m0.570s\n",
      "user\t0m0.039s\n",
      "sys\t0m0.027s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr4_7115521.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 7115523\n",
      "\t--012\n",
      "\t--out 1-Chr4-7115521_highConf\n",
      "\t--from-bp 7115519\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.151s\n",
      "user\t0m0.046s\n",
      "sys\t0m0.108s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr4_7115521.PL.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 7115523\n",
      "\t--012\n",
      "\t--out 1-Chr4-7115521_ALL\n",
      "\t--from-bp 7115519\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.214s\n",
      "user\t0m0.887s\n",
      "sys\t0m0.834s\n",
      "\n",
      "Chr4\t2\t4_2\t38397179\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.080s\n",
      "user\t0m0.041s\n",
      "sys\t0m0.023s\n",
      "\n",
      "real\t0m0.416s\n",
      "user\t0m0.074s\n",
      "sys\t0m0.021s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr4_38397179.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 38397181\n",
      "\t--012\n",
      "\t--out 2-Chr4-38397179_highConf\n",
      "\t--from-bp 38397177\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.156s\n",
      "user\t0m0.050s\n",
      "sys\t0m0.113s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr4_38397179.PL.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 38397181\n",
      "\t--012\n",
      "\t--out 2-Chr4-38397179_ALL\n",
      "\t--from-bp 38397177\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.676s\n",
      "user\t0m0.815s\n",
      "sys\t0m0.949s\n",
      "\n",
      "Chr4\t3\t4_3\t44460308\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.115s\n",
      "user\t0m0.078s\n",
      "sys\t0m0.017s\n",
      "\n",
      "real\t0m0.359s\n",
      "user\t0m0.104s\n",
      "sys\t0m0.014s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr4_44460308.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 44460310\n",
      "\t--012\n",
      "\t--out 3-Chr4-44460308_highConf\n",
      "\t--from-bp 44460306\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.157s\n",
      "user\t0m0.041s\n",
      "sys\t0m0.123s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr4_44460308.PL.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 44460310\n",
      "\t--012\n",
      "\t--out 3-Chr4-44460308_ALL\n",
      "\t--from-bp 44460306\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.463s\n",
      "user\t0m0.884s\n",
      "sys\t0m0.945s\n",
      "\n",
      "Chr4\t4\t4_4\t50972302\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.047s\n",
      "user\t0m0.023s\n",
      "sys\t0m0.020s\n",
      "\n",
      "real\t0m0.217s\n",
      "user\t0m0.029s\n",
      "sys\t0m0.026s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr4_50972302.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 50972304\n",
      "\t--012\n",
      "\t--out 4-Chr4-50972302_highConf\n",
      "\t--from-bp 50972300\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.163s\n",
      "user\t0m0.050s\n",
      "sys\t0m0.119s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr4_50972302.PL.vcf.gz\n",
      "\t--chr Chr4\n",
      "\t--to-bp 50972304\n",
      "\t--012\n",
      "\t--out 4-Chr4-50972302_ALL\n",
      "\t--from-bp 50972300\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.540s\n",
      "user\t0m0.889s\n",
      "sys\t0m0.983s\n"
     ]
    }
   ],
   "source": [
    "## Chr4\n",
    "chrom=Chr4\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in {1..4}\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1fe2b424-5411-4671-bd14-1e83f724481a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr6\t1\t6_1\t15125731\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.663s\n",
      "user\t0m0.032s\n",
      "sys\t0m0.014s\n",
      "\n",
      "real\t0m0.182s\n",
      "user\t0m0.025s\n",
      "sys\t0m0.017s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr6_15125731.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 15125733\n",
      "\t--012\n",
      "\t--out 1-Chr6-15125731_highConf\n",
      "\t--from-bp 15125729\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.169s\n",
      "user\t0m0.063s\n",
      "sys\t0m0.102s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr6_15125731.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 15125733\n",
      "\t--012\n",
      "\t--out 1-Chr6-15125731_ALL\n",
      "\t--from-bp 15125729\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.580s\n",
      "user\t0m0.849s\n",
      "sys\t0m0.842s\n",
      "\n",
      "Chr6\t2\t6_2\t32503688\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.199s\n",
      "user\t0m0.042s\n",
      "sys\t0m0.015s\n",
      "\n",
      "real\t0m0.067s\n",
      "user\t0m0.047s\n",
      "sys\t0m0.015s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr6_32503688.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 32503690\n",
      "\t--012\n",
      "\t--out 2-Chr6-32503688_highConf\n",
      "\t--from-bp 32503686\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.151s\n",
      "user\t0m0.033s\n",
      "sys\t0m0.123s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr6_32503688.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 32503690\n",
      "\t--012\n",
      "\t--out 2-Chr6-32503688_ALL\n",
      "\t--from-bp 32503686\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.025s\n",
      "user\t0m0.763s\n",
      "sys\t0m0.847s\n",
      "\n",
      "Chr6\t3\t6_3\t52921613\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.057s\n",
      "user\t0m0.013s\n",
      "sys\t0m0.022s\n",
      "\n",
      "real\t0m0.169s\n",
      "user\t0m0.020s\n",
      "sys\t0m0.021s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr6_52921613.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52921615\n",
      "\t--012\n",
      "\t--out 3-Chr6-52921613_highConf\n",
      "\t--from-bp 52921611\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.155s\n",
      "user\t0m0.044s\n",
      "sys\t0m0.119s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr6_52921613.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52921615\n",
      "\t--012\n",
      "\t--out 3-Chr6-52921613_ALL\n",
      "\t--from-bp 52921611\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m3.231s\n",
      "user\t0m0.812s\n",
      "sys\t0m1.004s\n",
      "\n",
      "Chr6\t4\t6_4\t52922047\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.037s\n",
      "user\t0m0.017s\n",
      "sys\t0m0.018s\n",
      "\n",
      "real\t0m0.040s\n",
      "user\t0m0.020s\n",
      "sys\t0m0.017s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr6_52922047.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52922049\n",
      "\t--012\n",
      "\t--out 4-Chr6-52922047_highConf\n",
      "\t--from-bp 52922045\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.156s\n",
      "user\t0m0.038s\n",
      "sys\t0m0.121s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr6_52922047.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52922049\n",
      "\t--012\n",
      "\t--out 4-Chr6-52922047_ALL\n",
      "\t--from-bp 52922045\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.241s\n",
      "user\t0m0.811s\n",
      "sys\t0m0.857s\n",
      "\n",
      "Chr6\t5\t6_5\t52966811\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.230s\n",
      "user\t0m0.068s\n",
      "sys\t0m0.020s\n",
      "\n",
      "real\t0m0.273s\n",
      "user\t0m0.082s\n",
      "sys\t0m0.027s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 5_Chr6_52966811.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52966813\n",
      "\t--012\n",
      "\t--out 5-Chr6-52966811_highConf\n",
      "\t--from-bp 52966809\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.174s\n",
      "user\t0m0.060s\n",
      "sys\t0m0.103s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 5_Chr6_52966811.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52966813\n",
      "\t--012\n",
      "\t--out 5-Chr6-52966811_ALL\n",
      "\t--from-bp 52966809\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.318s\n",
      "user\t0m0.881s\n",
      "sys\t0m0.906s\n",
      "\n",
      "Chr6\t6\t6_6\t52999112\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.261s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.036s\n",
      "\n",
      "real\t0m0.184s\n",
      "user\t0m0.150s\n",
      "sys\t0m0.028s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 6_Chr6_52999112.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52999114\n",
      "\t--012\n",
      "\t--out 6-Chr6-52999112_highConf\n",
      "\t--from-bp 52999110\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.199s\n",
      "user\t0m0.057s\n",
      "sys\t0m0.115s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 6_Chr6_52999112.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 52999114\n",
      "\t--012\n",
      "\t--out 6-Chr6-52999112_ALL\n",
      "\t--from-bp 52999110\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.511s\n",
      "user\t0m0.915s\n",
      "sys\t0m0.895s\n",
      "\n",
      "Chr6\t7\t6_7\t53016551\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.168s\n",
      "user\t0m0.089s\n",
      "sys\t0m0.029s\n",
      "\n",
      "real\t0m0.198s\n",
      "user\t0m0.148s\n",
      "sys\t0m0.036s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 7_Chr6_53016551.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53016553\n",
      "\t--012\n",
      "\t--out 7-Chr6-53016551_highConf\n",
      "\t--from-bp 53016549\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.149s\n",
      "user\t0m0.063s\n",
      "sys\t0m0.093s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 7_Chr6_53016551.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53016553\n",
      "\t--012\n",
      "\t--out 7-Chr6-53016551_ALL\n",
      "\t--from-bp 53016549\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.694s\n",
      "user\t0m0.914s\n",
      "sys\t0m0.968s\n",
      "\n",
      "Chr6\t8\t6_8\t53061080\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.177s\n",
      "user\t0m0.063s\n",
      "sys\t0m0.024s\n",
      "\n",
      "real\t0m0.389s\n",
      "user\t0m0.083s\n",
      "sys\t0m0.026s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 8_Chr6_53061080.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53061082\n",
      "\t--012\n",
      "\t--out 8-Chr6-53061080_highConf\n",
      "\t--from-bp 53061078\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.161s\n",
      "user\t0m0.042s\n",
      "sys\t0m0.118s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 8_Chr6_53061080.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53061082\n",
      "\t--012\n",
      "\t--out 8-Chr6-53061080_ALL\n",
      "\t--from-bp 53061078\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.715s\n",
      "user\t0m0.867s\n",
      "sys\t0m0.987s\n",
      "\n",
      "Chr6\t9\t6_9\t53083229\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.101s\n",
      "user\t0m0.085s\n",
      "sys\t0m0.012s\n",
      "\n",
      "real\t0m0.173s\n",
      "user\t0m0.120s\n",
      "sys\t0m0.029s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 9_Chr6_53083229.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53083231\n",
      "\t--012\n",
      "\t--out 9-Chr6-53083229_highConf\n",
      "\t--from-bp 53083227\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.157s\n",
      "user\t0m0.036s\n",
      "sys\t0m0.129s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 9_Chr6_53083229.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53083231\n",
      "\t--012\n",
      "\t--out 9-Chr6-53083229_ALL\n",
      "\t--from-bp 53083227\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.229s\n",
      "user\t0m0.857s\n",
      "sys\t0m0.913s\n",
      "\n",
      "Chr6\t10\t6_10\t53090212\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.083s\n",
      "user\t0m0.054s\n",
      "sys\t0m0.024s\n",
      "\n",
      "real\t0m0.098s\n",
      "user\t0m0.069s\n",
      "sys\t0m0.025s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 10_Chr6_53090212.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53090214\n",
      "\t--012\n",
      "\t--out 10-Chr6-53090212_highConf\n",
      "\t--from-bp 53090210\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.159s\n",
      "user\t0m0.060s\n",
      "sys\t0m0.105s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 10_Chr6_53090212.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53090214\n",
      "\t--012\n",
      "\t--out 10-Chr6-53090212_ALL\n",
      "\t--from-bp 53090210\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.220s\n",
      "user\t0m0.831s\n",
      "sys\t0m0.912s\n",
      "\n",
      "Chr6\t11\t6_11\t53094790\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.119s\n",
      "user\t0m0.087s\n",
      "sys\t0m0.017s\n",
      "\n",
      "real\t0m0.154s\n",
      "user\t0m0.134s\n",
      "sys\t0m0.013s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 11_Chr6_53094790.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53094792\n",
      "\t--012\n",
      "\t--out 11-Chr6-53094790_highConf\n",
      "\t--from-bp 53094788\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.155s\n",
      "user\t0m0.085s\n",
      "sys\t0m0.079s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 11_Chr6_53094790.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53094792\n",
      "\t--012\n",
      "\t--out 11-Chr6-53094790_ALL\n",
      "\t--from-bp 53094788\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.607s\n",
      "user\t0m0.928s\n",
      "sys\t0m0.879s\n",
      "\n",
      "Chr6\t12\t6_12\t53104387\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.059s\n",
      "user\t0m0.029s\n",
      "sys\t0m0.026s\n",
      "\n",
      "real\t0m0.072s\n",
      "user\t0m0.045s\n",
      "sys\t0m0.023s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 12_Chr6_53104387.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53104389\n",
      "\t--012\n",
      "\t--out 12-Chr6-53104387_highConf\n",
      "\t--from-bp 53104385\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.163s\n",
      "user\t0m0.062s\n",
      "sys\t0m0.107s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 12_Chr6_53104387.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53104389\n",
      "\t--012\n",
      "\t--out 12-Chr6-53104387_ALL\n",
      "\t--from-bp 53104385\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.499s\n",
      "user\t0m0.881s\n",
      "sys\t0m0.954s\n",
      "\n",
      "Chr6\t13\t6_13\t53195127\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.238s\n",
      "user\t0m0.084s\n",
      "sys\t0m0.017s\n",
      "\n",
      "real\t0m0.235s\n",
      "user\t0m0.101s\n",
      "sys\t0m0.033s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 13_Chr6_53195127.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53195129\n",
      "\t--012\n",
      "\t--out 13-Chr6-53195127_highConf\n",
      "\t--from-bp 53195125\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.162s\n",
      "user\t0m0.050s\n",
      "sys\t0m0.115s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 13_Chr6_53195127.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53195129\n",
      "\t--012\n",
      "\t--out 13-Chr6-53195127_ALL\n",
      "\t--from-bp 53195125\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.515s\n",
      "user\t0m0.951s\n",
      "sys\t0m0.911s\n",
      "\n",
      "Chr6\t14\t6_14\t53944185\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.392s\n",
      "user\t0m0.056s\n",
      "sys\t0m0.034s\n",
      "\n",
      "real\t0m0.185s\n",
      "user\t0m0.088s\n",
      "sys\t0m0.029s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 14_Chr6_53944185.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53944187\n",
      "\t--012\n",
      "\t--out 14-Chr6-53944185_highConf\n",
      "\t--from-bp 53944183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.149s\n",
      "user\t0m0.041s\n",
      "sys\t0m0.113s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 14_Chr6_53944185.PL.vcf.gz\n",
      "\t--chr Chr6\n",
      "\t--to-bp 53944187\n",
      "\t--012\n",
      "\t--out 14-Chr6-53944185_ALL\n",
      "\t--from-bp 53944183\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.126s\n",
      "user\t0m0.792s\n",
      "sys\t0m0.904s\n"
     ]
    }
   ],
   "source": [
    "## Chr6\n",
    "chrom=Chr6\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in {1..14}\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "584761f1-c003-403b-8f46-5268727d7502",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr7\t1\t7_1\t49828363\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.169s\n",
      "user\t0m0.082s\n",
      "sys\t0m0.015s\n",
      "\n",
      "real\t0m0.194s\n",
      "user\t0m0.078s\n",
      "sys\t0m0.021s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr7_49828363.vcf.gz\n",
      "\t--chr Chr7\n",
      "\t--to-bp 49828365\n",
      "\t--012\n",
      "\t--out 1-Chr7-49828363_highConf\n",
      "\t--from-bp 49828361\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.156s\n",
      "user\t0m0.045s\n",
      "sys\t0m0.116s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr7_49828363.PL.vcf.gz\n",
      "\t--chr Chr7\n",
      "\t--to-bp 49828365\n",
      "\t--012\n",
      "\t--out 1-Chr7-49828363_ALL\n",
      "\t--from-bp 49828361\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.350s\n",
      "user\t0m0.858s\n",
      "sys\t0m0.911s\n"
     ]
    }
   ],
   "source": [
    "## Chr7\n",
    "chrom=Chr7\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in 1\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cbb520d3-ea49-48e5-a297-9a503b0e4663",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr8\t1\t8_1\t27807304\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.327s\n",
      "user\t0m0.040s\n",
      "sys\t0m0.016s\n",
      "\n",
      "real\t0m0.250s\n",
      "user\t0m0.031s\n",
      "sys\t0m0.031s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr8_27807304.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 27807306\n",
      "\t--012\n",
      "\t--out 1-Chr8-27807304_highConf\n",
      "\t--from-bp 27807302\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.153s\n",
      "user\t0m0.032s\n",
      "sys\t0m0.125s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 1_Chr8_27807304.PL.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 27807306\n",
      "\t--012\n",
      "\t--out 1-Chr8-27807304_ALL\n",
      "\t--from-bp 27807302\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.929s\n",
      "user\t0m0.916s\n",
      "sys\t0m0.962s\n",
      "\n",
      "Chr8\t2\t8_2\t31864104\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.304s\n",
      "user\t0m0.061s\n",
      "sys\t0m0.020s\n",
      "\n",
      "real\t0m0.263s\n",
      "user\t0m0.088s\n",
      "sys\t0m0.024s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr8_31864104.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 31864106\n",
      "\t--012\n",
      "\t--out 2-Chr8-31864104_highConf\n",
      "\t--from-bp 31864102\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.156s\n",
      "user\t0m0.061s\n",
      "sys\t0m0.101s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 2_Chr8_31864104.PL.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 31864106\n",
      "\t--012\n",
      "\t--out 2-Chr8-31864104_ALL\n",
      "\t--from-bp 31864102\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.060s\n",
      "user\t0m0.782s\n",
      "sys\t0m0.932s\n",
      "\n",
      "Chr8\t3\t8_3\t38142268\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.077s\n",
      "user\t0m0.008s\n",
      "sys\t0m0.028s\n",
      "\n",
      "real\t0m0.054s\n",
      "user\t0m0.022s\n",
      "sys\t0m0.016s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr8_38142268.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 38142270\n",
      "\t--012\n",
      "\t--out 3-Chr8-38142268_highConf\n",
      "\t--from-bp 38142266\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.160s\n",
      "user\t0m0.049s\n",
      "sys\t0m0.117s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 3_Chr8_38142268.PL.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 38142270\n",
      "\t--012\n",
      "\t--out 3-Chr8-38142268_ALL\n",
      "\t--from-bp 38142266\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m3.593s\n",
      "user\t0m0.956s\n",
      "sys\t0m0.979s\n",
      "\n",
      "Chr8\t4\t8_4\t44933278\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.216s\n",
      "user\t0m0.070s\n",
      "sys\t0m0.015s\n",
      "\n",
      "real\t0m0.229s\n",
      "user\t0m0.079s\n",
      "sys\t0m0.028s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr8_44933278.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 44933280\n",
      "\t--012\n",
      "\t--out 4-Chr8-44933278_highConf\n",
      "\t--from-bp 44933276\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.161s\n",
      "user\t0m0.043s\n",
      "sys\t0m0.116s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 4_Chr8_44933278.PL.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 44933280\n",
      "\t--012\n",
      "\t--out 4-Chr8-44933278_ALL\n",
      "\t--from-bp 44933276\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.783s\n",
      "user\t0m0.917s\n",
      "sys\t0m1.021s\n",
      "\n",
      "Chr8\t5\t8_5\t53715058\n",
      "\n",
      "Subsetting VCF with bcftools\n",
      "\n",
      "real\t0m0.350s\n",
      "user\t0m0.060s\n",
      "sys\t0m0.025s\n",
      "\n",
      "real\t0m0.238s\n",
      "user\t0m0.080s\n",
      "sys\t0m0.022s\n",
      "Extracting stitch genotypes\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) bcftools/1.18 => bcftools/1.16     3) perl/5.36.0 => perl/5.34.0\n",
      "  2) htslib/1.18 => htslib/1.13         4) vcftools/0.1.16 => vcftools/20210912\n",
      "\n",
      "\n",
      "Indexing highConf-stitch VCF\n",
      "\n",
      "Get 012 values from stitch-highconfidence VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 5_Chr8_53715058.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 53715060\n",
      "\t--012\n",
      "\t--out 5-Chr8-53715058_highConf\n",
      "\t--from-bp 53715056\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "Extract all genotypes from STITCH\n",
      "\n",
      "real\t0m0.153s\n",
      "user\t0m0.035s\n",
      "sys\t0m0.122s\n",
      "\n",
      "Get 012 values from stitch-ALL VCF\n",
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf 5_Chr8_53715058.PL.vcf.gz\n",
      "\t--chr Chr8\n",
      "\t--to-bp 53715060\n",
      "\t--012\n",
      "\t--out 5-Chr8-53715058_ALL\n",
      "\t--from-bp 53715056\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GP,Number=3,Type=Float,Description=\"Posterior genotype probability of 0/0, 0/1, and 1/1\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
      "After filtering, kept 1074 out of 1074 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 1 out of a possible 1 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "Change longFormat sample names to shortFormat\n",
      "\n",
      "Create stitch GT file\n",
      "\n",
      "real\t0m2.369s\n",
      "user\t0m0.836s\n",
      "sys\t0m0.953s\n"
     ]
    }
   ],
   "source": [
    "## Chr8\n",
    "chrom=Chr8\n",
    "KASPcoords=~/snap_hap/impute/stitch_tests/coords/coords_for_stitch_longVersion_corrected_200kb.txt\n",
    "vcf=~/snap_hap/variants/stitch/${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "for snpID in {1..5}\n",
    "do\n",
    "    markerID=${chrom/Chr}_${snpID}\n",
    "    pos=$(grep -w $markerID $KASPcoords | cut -f8)\n",
    "    echo -e \"\\n\"$chrom\"\\t\"$snpID\"\\t\"$markerID\"\\t\"$pos\"\\n\"\n",
    "    # Subset VCF with bcftools\n",
    "    echo \"Subsetting VCF with bcftools\"\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.vcf.gz $vcf\n",
    "    time bcftools view -r ${chrom}:$pos -Oz -o ${snpID}_${chrom}_${pos}.PL.vcf.gz ${vcf/.vcf.gz/.PL.vcf.gz}    \n",
    "    # Extract genotypes\n",
    "    echo \"Extracting stitch genotypes\"\n",
    "    time bash ~/snap_hap/_scripts/bash/stitch/extract_stitchGT.sh $chrom $snpID $pos ${snpID}_${chrom}_${pos}.vcf.gz\n",
    "    rm ${snpID}_${chrom}_${pos}*.vcf.gz*\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "083e40f4-069a-4980-9ece-89cbe258cde7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract INFO, DEPTH and QUAL for each of these KASP markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4e68831b-5b6f-4217-8da3-f7629bb80e9d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1\t955185\tC\tA\t26533.4\tDP=2130;VDB=0.0403254;SGB=236.2;RPBZ=-4.67188;MQBZ=-3.6874;MQSBZ=-1.6169;BQBZ=0.448842;NMBZ=2.80605;SCBZ=-5.19645;FS=0;MQ0F=0.00892019;AC=853;AN=1246;DP4=388,266,563,415;MQ=42\n",
      "Chr1\t45535107\tC\tA\t28863.3\tDP=2464;VDB=0.00249496;SGB=270.867;RPBZ=0.0313099;MQBZ=-22.493;MQSBZ=-3.56976;BQBZ=-1.41371;NMBZ=22.3821;SCBZ=-0.146563;FS=0;MQ0F=0.000811688;AC=836;AN=1336;DP4=452,406,454,615;MQ=54\n",
      "Chr1\t47381180\tC\tA\t16317.9\tDP=2576;VDB=0.830319;SGB=237.663;RPBZ=2.0871;MQBZ=-12.9647;MQSBZ=-0.44103;BQBZ=0.708826;NMBZ=21.905;SCBZ=-0.106792;FS=0;MQ0F=0.00194099;AC=314;AN=1330;DP4=685,646,287,333;MQ=52\n",
      "Chr1\t51252783\tA\tC\t18432.5\tDP=2946;VDB=0.858512;SGB=259.479;RPBZ=0.499348;MQBZ=-16.429;MQSBZ=1.76715;BQBZ=0.737209;NMBZ=26.3499;SCBZ=0.468853;FS=0;MQ0F=0.00305499;AC=388;AN=1364;DP4=714,708,362,332;MQ=53\n",
      "Chr1\t60537705\tT\tC\t14255.6\tDP=2498;VDB=0.787776;SGB=291.111;RPBZ=0.954798;MQBZ=-9.90877;MQSBZ=3.08892;BQBZ=-2.68247;NMBZ=5.17687;SCBZ=4.84531;FS=0;MQ0F=0.0424339;AC=311;AN=1304;DP4=573,601,345,330;MQ=46\n",
      "Chr1\t71126303\tT\tG\t34834.4\tDP=2830;VDB=0.99977;SGB=401.123;RPBZ=-0.701631;MQBZ=-23.7337;MQSBZ=3.1658;BQBZ=0.830345;NMBZ=15.8517;SCBZ=0.0802863;FS=0;MQ0F=0.00353357;AC=909;AN=1432;DP4=539,405,723,514;MQ=54\n",
      "Chr2\t55473945\tA\tG\t22973.6\tDP=2175;VDB=0.955049;SGB=277.042;RPBZ=1.02531;MQBZ=-17.2302;MQSBZ=0.835852;BQBZ=-0.589419;NMBZ=-1.51153;SCBZ=-0.351318;FS=0;MQ0F=0.00183908;AC=445;AN=1324;DP4=450,516,377,437;MQ=57\n",
      "Chr2\t58751341\tA\tG\t8338.71\tDP=1981;VDB=0.00571555;SGB=173.942;RPBZ=5.04386;MQBZ=-15.7799;MQSBZ=-3.81457;BQBZ=9.62796;NMBZ=26.5719;SCBZ=21.1719;FS=0;MQ0F=0.04947;AC=221;AN=1110;DP4=511,487,130,262;MQ=49\n",
      "Chr3\t1055959\tT\tA\t25268.3\tDP=2470;VDB=0.855583;SGB=280.184;RPBZ=-0.358449;MQBZ=-1.14029;MQSBZ=0.487623;BQBZ=1.19652;NMBZ=-4.26463;SCBZ=0.0605525;FS=0;MQ0F=0.0105263;AC=813;AN=1316;DP4=470,377,588,457;MQ=41\n",
      "Chr3\t3488331\tC\tT\t29029.6\tDP=2497;VDB=0.791244;SGB=245.268;RPBZ=0.156619;MQBZ=-2.48471;MQSBZ=0.600968;BQBZ=1.11659;NMBZ=-7.71364;SCBZ=-0.353553;FS=0;MQ0F=0.000800961;AC=845;AN=1350;DP4=362,528,458,600;MQ=48\n",
      "Chr3\t44197469\tG\tC\t18392.9\tDP=2598;VDB=0.000425153;SGB=266.048;RPBZ=1.06044;MQBZ=9.26583;MQSBZ=-0.928831;BQBZ=-0.128311;NMBZ=-25.4601;SCBZ=-0.262521;FS=0;MQ0F=0.00153965;AC=368;AN=1368;DP4=761,502,439,251;MQ=50\n",
      "Chr4\t7115521\tT\tG\t32824.4\tDP=2682;VDB=0.316996;SGB=210.915;RPBZ=2.68549;MQBZ=1.28276;MQSBZ=2.03991;BQBZ=-4.22425;NMBZ=-13.6579;SCBZ=0.042898;FS=0;MQ0F=0.00149142;AC=948;AN=1408;DP4=379,557,438,688;MQ=48\n",
      "Chr4\t44460308\tA\tG\t26240.1\tDP=3136;VDB=0.999532;SGB=401.818;RPBZ=-0.202911;MQBZ=-6.49517;MQSBZ=1.30932;BQBZ=-0.111074;NMBZ=-13.1346;SCBZ=-0.360726;FS=0;MQ0F=0.00223214;AC=462;AN=1494;DP4=765,717,494,428;MQ=55\n",
      "Chr4\t50972302\tG\tA\t30206.3\tDP=2642;VDB=0.591301;SGB=268.314;RPBZ=-0.616136;MQBZ=-10.9921;MQSBZ=0.0504069;BQBZ=-1.62883;NMBZ=5.20243;SCBZ=0.872342;FS=0;MQ0F=0.00794852;AC=831;AN=1364;DP4=458,466,584,489;MQ=53\n",
      "Chr6\t15125731\tA\tT\t33795.9\tDP=2614;VDB=0.274234;SGB=297.331;RPBZ=0.645304;MQBZ=-23.8513;MQSBZ=2.2333;BQBZ=2.2461;NMBZ=18.8144;SCBZ=8.55267;FS=0;MQ0F=0.00229533;AC=965;AN=1440;DP4=405,368,663,577;MQ=53\n",
      "Chr6\t32503688\tA\tG\t25119.7\tDP=2748;VDB=0.365316;SGB=297.939;RPBZ=1.39349;MQBZ=-8.30178;MQSBZ=3.42896;BQBZ=-0.670485;NMBZ=-0.938114;SCBZ=1.18807;FS=0;MQ0F=0.0014556;AC=457;AN=1398;DP4=645,560,516,387;MQ=54\n",
      "Chr6\t52922047\tG\tT\t22605\tDP=2515;VDB=0.991365;SGB=252.517;RPBZ=0.524613;MQBZ=-6.53819;MQSBZ=1.02016;BQBZ=-0.153546;NMBZ=0.666596;SCBZ=-1.53605;FS=0;MQ0F=0.012326;AC=458;AN=1334;DP4=675,483,506,298;MQ=51\n",
      "Chr6\t52966811\tT\tA\t12147\tDP=1920;VDB=1.8251e-06;SGB=221.867;RPBZ=-1.92614;MQBZ=-3.48684;MQSBZ=-1.77815;BQBZ=-2.40476;NMBZ=-1.74185;SCBZ=1.81988;FS=0;MQ0F=0.0671875;AC=313;AN=1208;DP4=611,405,282,221;MQ=49\n",
      "Chr6\t52999112\tC\tT\t14710.1\tDP=2106;VDB=1;SGB=286.296;RPBZ=-1.15063;MQBZ=-19.194;MQSBZ=-1.34566;BQBZ=0.293295;NMBZ=27.9925;SCBZ=0.879052;FS=0;MQ0F=0.0014245;AC=346;AN=1200;DP4=533,481,270,324;MQ=50\n",
      "Chr6\t53016551\tC\tG\t18626.9\tDP=2535;VDB=0.623731;SGB=301.739;RPBZ=-0.265078;MQBZ=-13.144;MQSBZ=0.216468;BQBZ=-1.0142;NMBZ=-3.79224;SCBZ=-0.215135;FS=0;MQ0F=0.00118343;AC=318;AN=1364;DP4=669,649,325,321;MQ=57\n",
      "Chr6\t53061080\tA\tG\t22587.6\tDP=2113;VDB=0.156514;SGB=259.041;RPBZ=-2.76314;MQBZ=-15.4462;MQSBZ=0.942265;BQBZ=-0.152451;NMBZ=22.2517;SCBZ=2.91907;FS=0;MQ0F=0.00520587;AC=482;AN=1258;DP4=421,374,454,356;MQ=50\n",
      "Chr6\t53083229\tG\tA\t14730.8\tDP=2455;VDB=0.887193;SGB=322.291;RPBZ=-1.76965;MQBZ=-13.4262;MQSBZ=0.483341;BQBZ=-2.16984;NMBZ=-0.753788;SCBZ=0.185285;FS=0;MQ0F=0.00244399;AC=249;AN=1352;DP4=936,469,370,204;MQ=56\n",
      "Chr6\t53090212\tT\tC\t27083.1\tDP=2081;VDB=0.0654626;SGB=214.716;RPBZ=-0.475041;MQBZ=-20.2337;MQSBZ=2.50682;BQBZ=0.279777;NMBZ=10.5338;SCBZ=3.43208;FS=0;MQ0F=0.00576646;AC=870;AN=1338;DP4=400,324,612,349;MQ=55\n",
      "Chr6\t53094790\tC\tA\t14456.7\tDP=2600;VDB=0.269369;SGB=350.434;RPBZ=-1.56544;MQBZ=-4.93324;MQSBZ=-1.57145;BQBZ=-0.0860416;NMBZ=-3.97467;SCBZ=-0.00329092;FS=0;MQ0F=0.00269231;AC=239;AN=1396;DP4=863,616,319,234;MQ=55\n",
      "Chr6\t53104387\tA\tC\t19319\tDP=2586;VDB=0.829882;SGB=253.813;RPBZ=0.17564;MQBZ=-25.4735;MQSBZ=-0.57573;BQBZ=-2.02122;NMBZ=34.3926;SCBZ=1.23405;FS=0;MQ0F=0.00116009;AC=433;AN=1358;DP4=544,593,347,399;MQ=53\n",
      "Chr6\t53195127\tC\tT\t41505.6\tDP=2618;VDB=0.989784;SGB=252.215;RPBZ=-1.26149;MQBZ=-20.2318;MQSBZ=-0.597156;BQBZ=2.41543;NMBZ=7.03615;SCBZ=-0.523486;FS=0;MQ0F=0.00229183;AC=1089;AN=1410;DP4=417,261,756,589;MQ=55\n",
      "Chr6\t53944185\tG\tA\t11483.1\tDP=2610;VDB=0.237318;SGB=285.538;RPBZ=2.17298;MQBZ=-7.01265;MQSBZ=-2.1303;BQBZ=-1.02699;NMBZ=-6.65154;SCBZ=-0.156828;FS=0;MQ0F=0.00229885;AC=197;AN=1400;DP4=715,882,194,267;MQ=52\n",
      "Chr7\t49828363\tT\tA\t18419.7\tDP=3023;VDB=0.947225;SGB=265.743;RPBZ=-0.914265;MQBZ=-13.2196;MQSBZ=0.467364;BQBZ=-3.10809;NMBZ=17.5788;SCBZ=5.40965;FS=0;MQ0F=0.00793913;AC=426;AN=1434;DP4=804,736,453,308;MQ=48\n",
      "Chr8\t27807304\tT\tG\t17404.8\tDP=2532;VDB=0.970465;SGB=206.139;RPBZ=-0.850395;MQBZ=-7.93494;MQSBZ=1.05989;BQBZ=-0.856474;NMBZ=-1.21687;SCBZ=-0.00943756;FS=0;MQ0F=0.00276461;AC=381;AN=1390;DP4=737,611,337,292;MQ=50\n",
      "Chr8\t38142268\tC\tG\t27897.2\tDP=2567;VDB=0.304821;SGB=245.05;RPBZ=0.61162;MQBZ=-20.6605;MQSBZ=-1.27325;BQBZ=-0.549615;NMBZ=1.38198;SCBZ=-0.27221;FS=0;MQ0F=0.0019478;AC=546;AN=1376;DP4=395,599,394,584;MQ=56\n"
     ]
    }
   ],
   "source": [
    "## INFO from mpileup output\n",
    "\n",
    "cd ~/snap_hap/variants/vcf_bcftools_Am_all/\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    bcftools view -H -R ../stitch/KASP_validation.pos ${chrom}/Am_all_bcftools_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz | cut -f1-2,4-6,8\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "624c68c8-a197-4525-b12f-c93c9ea08e60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1\t955185\tC\tA\tEAF=0.61127;INFO_SCORE=0.69969;HWE=0.000312;ERC=1044.39;EAC=578.828;PAF=0.64341\n",
      "Chr1\t45535107\tC\tA\tEAF=0.53496;INFO_SCORE=0.8312;HWE=0.625;ERC=1051.56;EAC=862.859;PAF=0.54928\n",
      "Chr1\t47381180\tC\tA\tEAF=0.29956;INFO_SCORE=0.72775;HWE=0.236;ERC=658.886;EAC=1351.61;PAF=0.32772\n",
      "Chr1\t51252783\tA\tC\tEAF=0.36975;INFO_SCORE=0.80042;HWE=0.503;ERC=809.807;EAC=1355.39;PAF=0.37401\n",
      "Chr1\t60537705\tT\tC\tEAF=0.34552;INFO_SCORE=0.5466;HWE=0.173;ERC=613.451;EAC=1163.09;PAF=0.34531\n",
      "Chr1\t71126303\tT\tG\tEAF=0.53355;INFO_SCORE=0.79213;HWE=0.854;ERC=1221.66;EAC=954.728;PAF=0.56133\n",
      "Chr2\t55473945\tA\tG\tEAF=0.4309;INFO_SCORE=0.7368;HWE=0.9;ERC=787.793;EAC=947.772;PAF=0.45391\n",
      "Chr2\t58751341\tA\tG\tEAF=0.38488;INFO_SCORE=0.61493;HWE=0.655;ERC=396.784;EAC=977.689;PAF=0.28868\n",
      "Chr3\t1055959\tT\tA\tEAF=0.53935;INFO_SCORE=0.37925;HWE=5.79e-06;ERC=1016.29;EAC=841.455;PAF=0.54706\n",
      "Chr3\t3488331\tC\tT\tEAF=0.56796;INFO_SCORE=0.61748;HWE=0.448;ERC=1122.85;EAC=928.436;PAF=0.54739\n",
      "Chr3\t44197469\tG\tC\tEAF=0.37939;INFO_SCORE=0.76898;HWE=0.692;ERC=710.888;EAC=1356.5;PAF=0.34386\n",
      "Chr4\t7115521\tT\tG\tEAF=0.6432;INFO_SCORE=0.82656;HWE=0.424;ERC=1236.62;EAC=953.787;PAF=0.56456\n",
      "Chr4\t44460308\tA\tG\tEAF=0.38262;INFO_SCORE=0.66304;HWE=0.649;ERC=969.835;EAC=1486.49;PAF=0.39483\n",
      "Chr4\t50972302\tG\tA\tEAF=0.52154;INFO_SCORE=0.59632;HWE=0.0659;ERC=1126.53;EAC=976.853;PAF=0.53558\n",
      "Chr6\t15125731\tA\tT\tEAF=0.58775;INFO_SCORE=0.89667;HWE=0.0886;ERC=1235.17;EAC=775.926;PAF=0.61418\n",
      "Chr6\t32503688\tA\tG\tEAF=0.41623;INFO_SCORE=0.87169;HWE=0.489;ERC=920.796;EAC=1259.66;PAF=0.4223\n",
      "Chr6\t52922047\tG\tT\tEAF=0.42833;INFO_SCORE=0.60069;HWE=0.901;ERC=867.785;EAC=1073.66;PAF=0.44698\n",
      "Chr6\t52966811\tT\tA\tEAF=0.32062;INFO_SCORE=0.59027;HWE=2.44e-05;ERC=491.794;EAC=917.532;PAF=0.34896\n",
      "Chr6\t52999112\tC\tT\tEAF=0.39177;INFO_SCORE=0.80443;HWE=3.42e-20;ERC=673.702;EAC=1025.7;PAF=0.39643\n",
      "Chr6\t53016551\tC\tG\tEAF=0.31391;INFO_SCORE=0.81158;HWE=1.28e-12;ERC=674.91;EAC=1382.58;PAF=0.32803\n",
      "Chr6\t53061080\tA\tG\tEAF=0.48159;INFO_SCORE=0.71482;HWE=1.81e-13;ERC=862.369;EAC=818.918;PAF=0.51292\n",
      "Chr6\t53083229\tG\tA\tEAF=0.22171;INFO_SCORE=0.61022;HWE=9.79e-07;ERC=531.921;EAC=1350.58;PAF=0.28256\n",
      "Chr6\t53090212\tT\tC\tEAF=0.55551;INFO_SCORE=0.53495;HWE=0.179;ERC=926.636;EAC=748.844;PAF=0.55306\n",
      "Chr6\t53094790\tC\tA\tEAF=0.23235;INFO_SCORE=0.79063;HWE=2.17e-05;ERC=531.993;EAC=1543.47;PAF=0.25632\n",
      "Chr6\t53104387\tA\tC\tEAF=0.39253;INFO_SCORE=0.69188;HWE=0.00945;ERC=767.775;EAC=1179.61;PAF=0.39426\n",
      "Chr6\t53195127\tC\tT\tEAF=0.68807;INFO_SCORE=0.4721;HWE=0.54;ERC=1401.47;EAC=673.861;PAF=0.6753\n",
      "Chr6\t53944185\tG\tA\tEAF=0.1986;INFO_SCORE=0.7492;HWE=0.135;ERC=460.049;EAC=1570.45;PAF=0.22657\n",
      "Chr7\t49828363\tT\tA\tEAF=0.31702;INFO_SCORE=0.59881;HWE=0.228;ERC=810.614;EAC=1523.43;PAF=0.3473\n",
      "Chr8\t27807304\tT\tG\tEAF=0.31806;INFO_SCORE=0.80568;HWE=0.307;ERC=696.932;EAC=1288.48;PAF=0.35103\n",
      "Chr8\t38142268\tC\tG\tEAF=0.51054;INFO_SCORE=0.67239;HWE=0.501;ERC=1029.6;EAC=1035.76;PAF=0.49851\n"
     ]
    }
   ],
   "source": [
    "## INFO from stitch finalRun output\n",
    "\n",
    "cd ~/snap_hap/variants/stitch\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    bcftools view -H -R KASP_validation.pos ${chrom}/*.PL.vcf.gz | cut -f1,2,4,5,8\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "251536b5-ab79-45ba-a74d-6d4a6505c30b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1\t955185\tC\tA\tEAF=0.61852;INFO_SCORE=0.71681;HWE=6.78e-05;ERC=1069.38;EAC=598.826;PAF=0.64104\n",
      "Chr1\t45535107\tC\tA\tEAF=0.5282;INFO_SCORE=0.7465;HWE=0.143;ERC=1094.52;EAC=899.864;PAF=0.5488\n",
      "Chr1\t47381180\tC\tA\tEAF=0.31531;INFO_SCORE=0.62547;HWE=0.0201;ERC=672.887;EAC=1400.6;PAF=0.32452\n",
      "Chr1\t51252783\tA\tC\tEAF=0.36165;INFO_SCORE=0.72718;HWE=0.841;ERC=824.806;EAC=1377.38;PAF=0.37454\n",
      "Chr1\t60537705\tT\tC\tEAF=0.34305;INFO_SCORE=0.58476;HWE=0.685;ERC=633.438;EAC=1188.07;PAF=0.34775\n",
      "Chr1\t71126303\tT\tG\tEAF=0.54114;INFO_SCORE=0.72912;HWE=0.157;ERC=1273.65;EAC=982.725;PAF=0.56447\n",
      "Chr2\t55473945\tA\tG\tEAF=0.42485;INFO_SCORE=0.76758;HWE=0.851;ERC=824.781;EAC=991.766;PAF=0.45404\n",
      "Chr2\t58751341\tA\tG\tEAF=0.34912;INFO_SCORE=0.52485;HWE=0.351;ERC=431.774;EAC=1014.69;PAF=0.2985\n",
      "Chr3\t1055959\tT\tA\tEAF=0.53891;INFO_SCORE=0.42271;HWE=0.0111;ERC=1037.29;EAC=871.45;PAF=0.54344\n",
      "Chr3\t3488331\tC\tT\tEAF=0.54822;INFO_SCORE=0.61503;HWE=0.118;ERC=1132.85;EAC=952.426;PAF=0.54326\n",
      "Chr3\t44197469\tG\tC\tEAF=0.37404;INFO_SCORE=0.7342;HWE=0.507;ERC=736.885;EAC=1424.48;PAF=0.34093\n",
      "Chr4\t7115521\tT\tG\tEAF=0.63899;INFO_SCORE=0.71536;HWE=0.505;ERC=1272.62;EAC=972.786;PAF=0.56677\n",
      "Chr4\t44460308\tA\tG\tEAF=0.37706;INFO_SCORE=0.66146;HWE=0.645;ERC=1009.83;EAC=1548.47;PAF=0.39473\n",
      "Chr4\t50972302\tG\tA\tEAF=0.52218;INFO_SCORE=0.6085;HWE=0.244;ERC=1131.54;EAC=976.852;PAF=0.53668\n",
      "Chr6\t53061080\tA\tG\tEAF=0.47829;INFO_SCORE=0.73748;HWE=1.07e-14;ERC=871.373;EAC=840.91;PAF=0.5089\n",
      "Chr6\t53083229\tG\tA\tEAF=0.2255;INFO_SCORE=0.63891;HWE=1.19e-10;ERC=534.927;EAC=1389.56;PAF=0.27796\n",
      "Chr6\t53090212\tT\tC\tEAF=0.56283;INFO_SCORE=0.53206;HWE=0.713;ERC=952.633;EAC=765.84;PAF=0.55435\n",
      "Chr6\t53094790\tC\tA\tEAF=0.24184;INFO_SCORE=0.7273;HWE=4.13e-07;ERC=547.994;EAC=1590.45;PAF=0.25626\n",
      "Chr6\t53104387\tA\tC\tEAF=0.39604;INFO_SCORE=0.66533;HWE=0.0122;ERC=786.774;EAC=1196.6;PAF=0.39669\n",
      "Chr6\t53061080\tA\tG\tEAF=0.46806;INFO_SCORE=0.72134;HWE=2.13e-13;ERC=873.367;EAC=838.913;PAF=0.51006\n",
      "Chr6\t53083229\tG\tA\tEAF=0.23408;INFO_SCORE=0.67904;HWE=2.98e-08;ERC=534.927;EAC=1387.56;PAF=0.27825\n",
      "Chr6\t53090212\tT\tC\tEAF=0.55192;INFO_SCORE=0.56116;HWE=0.582;ERC=950.633;EAC=764.84;PAF=0.55415\n",
      "Chr6\t53094790\tC\tA\tEAF=0.24816;INFO_SCORE=0.75962;HWE=1.44e-07;ERC=547.994;EAC=1584.45;PAF=0.25698\n",
      "Chr6\t53104387\tA\tC\tEAF=0.40307;INFO_SCORE=0.68729;HWE=0.0213;ERC=786.774;EAC=1194.6;PAF=0.39709\n",
      "Chr6\t53061080\tA\tG\tEAF=0.48087;INFO_SCORE=0.72805;HWE=2.75e-12;ERC=895.358;EAC=868.903;PAF=0.5075\n",
      "Chr6\t53083229\tG\tA\tEAF=0.22995;INFO_SCORE=0.62714;HWE=4.26e-05;ERC=554.932;EAC=1433.53;PAF=0.27908\n",
      "Chr6\t53090212\tT\tC\tEAF=0.56279;INFO_SCORE=0.56534;HWE=0.243;ERC=979.624;EAC=777.835;PAF=0.55741\n",
      "Chr6\t53094790\tC\tA\tEAF=0.23498;INFO_SCORE=0.78314;HWE=1.07e-07;ERC=557.997;EAC=1627.43;PAF=0.25533\n",
      "Chr6\t53104387\tA\tC\tEAF=0.39608;INFO_SCORE=0.65289;HWE=0.000618;ERC=802.781;EAC=1240.57;PAF=0.39287\n",
      "Chr6\t53195127\tC\tT\tEAF=0.68265;INFO_SCORE=0.53762;HWE=0.309;ERC=1503.43;EAC=705.857;PAF=0.68051\n",
      "Chr6\t53944185\tG\tA\tEAF=0.20263;INFO_SCORE=0.67289;HWE=0.0923;ERC=483.056;EAC=1689.41;PAF=0.22235\n",
      "Chr6\t15125731\tA\tT\tEAF=0.58887;INFO_SCORE=0.76969;HWE=0.571;ERC=1252.16;EAC=790.923;PAF=0.61288\n",
      "Chr6\t32503688\tA\tG\tEAF=0.41814;INFO_SCORE=0.80637;HWE=0.801;ERC=951.793;EAC=1309.64;PAF=0.42088\n",
      "Chr6\t52922047\tG\tT\tEAF=0.42854;INFO_SCORE=0.58307;HWE=0.709;ERC=901.773;EAC=1112.65;PAF=0.44766\n",
      "Chr6\t52966811\tT\tA\tEAF=0.31322;INFO_SCORE=0.57991;HWE=5.78e-06;ERC=513.786;EAC=953.526;PAF=0.35015\n",
      "Chr6\t52922047\tG\tT\tEAF=0.42374;INFO_SCORE=0.61915;HWE=0.458;ERC=899.774;EAC=1110.65;PAF=0.44755\n",
      "Chr6\t52966811\tT\tA\tEAF=0.30093;INFO_SCORE=0.57729;HWE=0.00512;ERC=513.786;EAC=953.526;PAF=0.35015\n",
      "Chr6\t52922047\tG\tT\tEAF=0.43447;INFO_SCORE=0.64695;HWE=0.0104;ERC=899.775;EAC=1115.65;PAF=0.44644\n",
      "Chr6\t52966811\tT\tA\tEAF=0.30828;INFO_SCORE=0.66708;HWE=3.33e-08;ERC=515.785;EAC=966.521;PAF=0.34796\n",
      "Chr6\t52999112\tC\tT\tEAF=0.38385;INFO_SCORE=0.83148;HWE=3.53e-20;ERC=692.704;EAC=1074.68;PAF=0.39194\n",
      "Chr6\t53016551\tC\tG\tEAF=0.32384;INFO_SCORE=0.73648;HWE=2.55e-11;ERC=714.907;EAC=1465.56;PAF=0.32787\n",
      "Chr6\t52966811\tT\tA\tEAF=0.34618;INFO_SCORE=0.67324;HWE=3.06e-10;ERC=521.787;EAC=964.503;PAF=0.35107\n",
      "Chr6\t52999112\tC\tT\tEAF=0.37899;INFO_SCORE=0.76275;HWE=3.12e-22;ERC=693.704;EAC=1073.68;PAF=0.3925\n",
      "Chr6\t53016551\tC\tG\tEAF=0.3286;INFO_SCORE=0.78592;HWE=1.39e-14;ERC=705.908;EAC=1453.56;PAF=0.32689\n",
      "Chr6\t52966811\tT\tA\tEAF=0.32869;INFO_SCORE=0.51052;HWE=0.0428;ERC=496.793;EAC=936.516;PAF=0.34661\n",
      "Chr6\t52999112\tC\tT\tEAF=0.37977;INFO_SCORE=0.85462;HWE=1e-19;ERC=676.699;EAC=1031.7;PAF=0.3961\n",
      "Chr6\t53016551\tC\tG\tEAF=0.32569;INFO_SCORE=0.75283;HWE=1.71e-14;ERC=692.903;EAC=1395.58;PAF=0.33177\n",
      "Chr6\t53061080\tA\tG\tEAF=0.50333;INFO_SCORE=0.73336;HWE=1.26e-12;ERC=885.36;EAC=855.909;PAF=0.50846\n",
      "Chr6\t53016551\tC\tG\tEAF=0.32003;INFO_SCORE=0.86596;HWE=8.76e-14;ERC=687.904;EAC=1383.58;PAF=0.33208\n",
      "Chr6\t53061080\tA\tG\tEAF=0.49063;INFO_SCORE=0.66725;HWE=1.05e-14;ERC=880.365;EAC=841.913;PAF=0.51116\n",
      "Chr6\t53083229\tG\tA\tEAF=0.21946;INFO_SCORE=0.62912;HWE=1.4e-06;ERC=535.928;EAC=1396.56;PAF=0.27733\n",
      "Chr6\t53090212\tT\tC\tEAF=0.54529;INFO_SCORE=0.55224;HWE=0.175;ERC=952.63;EAC=768.84;PAF=0.55338\n",
      "Chr6\t53094790\tC\tA\tEAF=0.23298;INFO_SCORE=0.72796;HWE=1.43e-06;ERC=547.995;EAC=1594.45;PAF=0.25578\n",
      "Chr6\t53104387\tA\tC\tEAF=0.41596;INFO_SCORE=0.66143;HWE=0.0683;ERC=801.767;EAC=1221.59;PAF=0.39626\n",
      "Chr6\t53061080\tA\tG\tEAF=0.47901;INFO_SCORE=0.70637;HWE=5.23e-16;ERC=872.367;EAC=837.913;PAF=0.51007\n",
      "Chr6\t53083229\tG\tA\tEAF=0.22607;INFO_SCORE=0.6225;HWE=4.94e-05;ERC=534.927;EAC=1388.56;PAF=0.2781\n",
      "Chr6\t53090212\tT\tC\tEAF=0.55976;INFO_SCORE=0.56115;HWE=0.806;ERC=949.633;EAC=765.84;PAF=0.55357\n",
      "Chr6\t53094790\tC\tA\tEAF=0.22933;INFO_SCORE=0.78499;HWE=7.09e-08;ERC=547.994;EAC=1588.45;PAF=0.2565\n",
      "Chr6\t53104387\tA\tC\tEAF=0.38993;INFO_SCORE=0.63795;HWE=0.00175;ERC=786.774;EAC=1196.6;PAF=0.39669\n",
      "Chr7\t49828363\tT\tA\tEAF=0.35735;INFO_SCORE=0.62015;HWE=0.0597;ERC=820.608;EAC=1560.42;PAF=0.34464\n",
      "Chr8\t27807304\tT\tG\tEAF=0.31018;INFO_SCORE=0.73593;HWE=0.167;ERC=718.936;EAC=1334.45;PAF=0.35012\n",
      "Chr8\t38142268\tC\tG\tEAF=0.52933;INFO_SCORE=0.67069;HWE=0.713;ERC=1058.58;EAC=1050.75;PAF=0.50186\n"
     ]
    }
   ],
   "source": [
    "## INFO from stitch validation output\n",
    "\n",
    "cd ~/snap_hap/impute/stitch_tests\n",
    "for vcf in */run3.1*/run3_*-chr?_*K75_cov20*/*.PL.vcf.gz\n",
    "do\n",
    "    # echo $vcf\n",
    "    # bash ~/snap_hap/_scripts/bash/stitch/add_PG_PL.sh $vcf\n",
    "    bcftools view -H -R ~/snap_hap/variants/stitch/KASP_validation.pos $vcf | cut -f1,2,4,5,8\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1f460ad-1682-49fc-9891-345c38fef1ed",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract stitch INFO, freqs, genotypes and  positions in all 30 KASP positions + 100kb around. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "de1f0a97-43fc-4288-9b5d-732ad8327a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract stitch INFO, params, GTs final Runs\n",
    "cd ~/snap_hap/variants/stitch\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    vcf=${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.vcf.gz\n",
    "    bcftools query -R ./INFO_acrossRuns/KASP_validation_100kb.pos --format \"%CHROM\\t%POS\\t%EAF\\t%INFO_SCORE\\t%HWE\\t%ERC\\t%EAC\\t%PAF[\\t%GT]\\n\" $vcf >> ~/snap_hap/variants/stitch/INFO_acrossRuns/stitch.final.params\n",
    "done\n",
    "\n",
    "## Extract stitch INFO, params, GTs validation runs\n",
    "cd ~/snap_hap/impute/stitch_tests\n",
    "for vcf in ./Chr?/run3.1*/run3_*-chr?_*K75_cov20*/*.PL.vcf.gz\n",
    "do\n",
    "    bcftools query -R ~/snap_hap/variants/stitch/INFO_acrossRuns/KASP_validation_100kb.pos --format \"%CHROM\\t%POS\\t%EAF\\t%INFO_SCORE\\t%HWE\\t%ERC\\t%EAC\\t%PAF[\\t%GT]\\n\" $vcf >> ~/snap_hap/variants/stitch/INFO_acrossRuns/stitch.validation.params\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "41fdf4bc-db86-4387-8299-62d60d08fac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "142836 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/INFO_acrossRuns/stitch.final.params\n",
      "197169 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/INFO_acrossRuns/stitch.validation.params\n"
     ]
    }
   ],
   "source": [
    "wc -l ~/snap_hap/variants/stitch/INFO_acrossRuns/stitch.final.params\n",
    "wc -l ~/snap_hap/variants/stitch/INFO_acrossRuns/stitch.validation.params"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e248a66-ee23-448d-97c6-600088b01937",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Getting final STITCH parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a88fe188-6b20-4139-b15f-3fd128d976ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/snap_hap/variants/stitch/\n",
    "\n",
    "for chrom in Chr{1..8}\n",
    "do    \n",
    "    time bcftools query --format \"%CHROM\\t%POS\\t%EAF\\t%INFO_SCORE\\t%HWE\\t%ERC\\t%EAC\\t%PAF\" ${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.vcf.gz > ${chrom}/Am_all_stitch_${chrom}.params\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb887c9-5d22-472f-aa26-5383b5c0922d",
   "metadata": {},
   "source": [
    "## Extract allele frequencies and counts for stitched VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e053043e-e153-4b8f-ab3d-9620ea2829c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch\n",
    "\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    sbatch -J stat${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-statVcfs.sbatch.sh ${chrom}/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.vcf.gz ${chrom}/Am_all_stitch_${chrom}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb695704-8320-4611-b97c-e99b852bb9db",
   "metadata": {},
   "source": [
    "## Remove invariant sites from final VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02042ff1-bee2-429e-8f27-3df28f489ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "stitchRun=stitch\n",
    "baseDIR=/nfs/scistore18/bartogrp/apal/snap_hap/variants/$stitchRun\n",
    "outPrefixRun=stitchRun1\n",
    "\n",
    "cd $baseDIR\n",
    "for chrom in Chr6\n",
    "do\n",
    "    inVCF=$baseDIR/$chrom/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.tagged.vcf.gz\n",
    "    outVCF=$baseDIR/$chrom/Am_all_${outPrefixRun}_${chrom}.final.vcf.gz\n",
    "    time bcftools view -e 'AC==0 | AC==AN' -Oz -o $outVCF $inVCF\n",
    "    time bcftools tabix -f $outVCF\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a5a193-ec52-4bd5-a2eb-54cfec321c97",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Cluster implmentation\n",
    "stitchRun=stitch ## Run1\n",
    "baseDIR=/nfs/scistore18/bartogrp/apal/snap_hap/variants/$stitchRun\n",
    "outPrefixRun=stitchRun1\n",
    "\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/postSTITCH_run1/removeSites\n",
    "# NB. array information is the chromosome no.\n",
    "sbatch -J removeInvariantSites --array=1-8 ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-remove_InvariantSites.sbatch.sh $stitchRun $outPrefixRun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5981927-034b-422b-9084-8075c2242415",
   "metadata": {},
   "outputs": [],
   "source": [
    "## No. of sites for stitch Run 1\n",
    "stitchRun=stitch\n",
    "baseDIR=/nfs/scistore18/bartogrp/apal/snap_hap/variants/$stitchRun\n",
    "\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    vcf=$baseDIR/$chrom/*final.vcf.gz\n",
    "    echo $chrom $vcf\n",
    "    zgrep -c ^Chr $vcf\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d375fe8-b3e9-43e0-ab43-fbc03d626268",
   "metadata": {},
   "source": [
    "Chr1 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr1/Am_all_stitchRun1_Chr1.final.vcf.gz\n",
    "2940701\n",
    "\n",
    "Chr2 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr2/Am_all_stitchRun1_Chr2.final.vcf.gz\n",
    "3301920\n",
    "\n",
    "Chr3 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr3/Am_all_stitchRun1_Chr3.final.vcf.gz\n",
    "2600018\n",
    "\n",
    "Chr4 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr4/Am_all_stitchRun1_Chr4.final.vcf.gz\n",
    "2317799\n",
    "\n",
    "Chr5 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr5/Am_all_stitchRun1_Chr5.final.vcf.gz\n",
    "3277515\n",
    "\n",
    "Chr6 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr6/Am_all_stitchRun1_Chr6.final.vcf.gz\n",
    "2533823\n",
    "\n",
    "Chr7 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr7/Am_all_stitchRun1_Chr7.final.vcf.gz\n",
    "2358818\n",
    "\n",
    "Chr8 /nfs/scistore18/bartogrp/apal/snap_hap/variants/stitch/Chr8/Am_all_stitchRun1_Chr8.final.vcf.gz\n",
    "2765149"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e643e061-faaf-4d04-8a50-57842b912886",
   "metadata": {},
   "source": [
    "# Postprocessing stitch Run2 and Run3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0c3350f-1d28-4a2d-8e67-60578e00065b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Always initiate this cell!!!\n",
    "stitchRun=stitch_run2\n",
    "run=run2\n",
    "# stitchRun=stitch_run3 ### Run3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57dcb307-c272-4d3a-b69a-495397c15456",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make a list of stitch highConfidence chromSegment files\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    baseDIR=~/snap_hap/variants/$stitchRun/${chrom}\n",
    "    echo -e $chrom'\\t'$baseDIR'\\t'$(realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz | wc -l) files\n",
    "    realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*[^PL].vcf.gz > ${baseDIR}/stitchVcfs.highConf.list\n",
    "    echo -e ${baseDIR}/stitchVcfs.highConf.list'\\n'\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68b31d1-6718-4ba4-aaf6-bc69a8c2d450",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract all GTs from stitch highConfidence output\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/postSTITCH_$run/extractGT\n",
    "\n",
    "# Chr1 DONE\n",
    "chrom=Chr1\n",
    "sbatch --array=1-72 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "# Chr2 DONE\n",
    "chrom=Chr2\n",
    "sbatch --array=1-78 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "# Chr3\n",
    "chrom=Chr3\n",
    "sbatch --array=1-65 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "\n",
    "# Chr4\n",
    "chrom=Chr4\n",
    "sbatch --array=1-55 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "# Chr5\n",
    "chrom=Chr5\n",
    "sbatch --array=1-71 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "\n",
    "# Chr6 DONE\n",
    "chrom=Chr6\n",
    "sbatch --array=1-56 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "# Chr7\n",
    "chrom=Chr7\n",
    "sbatch --array=1-56 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list\n",
    "\n",
    "\n",
    "# Chr8\n",
    "chrom=Chr8\n",
    "sbatch --array=1-54 -J ${chrom}_postSTITCH-extractGT ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-extractGT.sbatch.sh \\\n",
    "    $chrom $stitchRun ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.highConf.list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d663f91-32e6-48ac-9711-897002209ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make a list of stitch all chromSegment files\n",
    "for chrom in Chr1 #done: Chr6\n",
    "do\n",
    "    baseDIR=~/snap_hap/variants/$stitchRun/${chrom}\n",
    "    echo -e $chrom'\\t'$baseDIR'\\t'$(realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*.PL.vcf.gz | wc -l) files\n",
    "    realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*.PL.vcf.gz > ${baseDIR}/stitchVcfs.ALL.list\n",
    "    echo -e ${baseDIR}/stitchVcfs.ALL.list'\\n'\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c0b2e1-7ed6-4262-917b-664892437b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add AC,AN annotations to stitch ALL vcf file\n",
    "for chrom in Chr1 #done: Chr6\n",
    "do\n",
    "    while read inputVCF\n",
    "    do\n",
    "        echo $inputVCF\n",
    "        time bash ~/snap_hap/_scripts/bash/fill-AC-AN_vcfgzFormat.sh $inputVCF\n",
    "    done < ~/snap_hap/variants/$stitchRun/$chrom/stitchVcfs.ALL.list\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af88a8e-b71a-4951-a36d-bcb039df6acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make a list of tagged (AC,AN) stitch all chromSegment files\n",
    "for chrom in Chr{1..8}\n",
    "do\n",
    "    baseDIR=~/snap_hap/variants/$stitchRun/${chrom}\n",
    "    echo -e $chrom'\\t'$baseDIR'\\t'$(realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*.PL.tagged.vcf.gz | wc -l) files\n",
    "    realpath $baseDIR/stitch_chromSegments/*/stitch.${chrom}.*.*.PL.tagged.vcf.gz > ${baseDIR}/stitchVcfs.ALL.tagged.list\n",
    "    echo -e ${baseDIR}/stitchVcfs.highConf.list'\\n'\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f62698-5cbe-43c4-91ee-1fb6d1e29870",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### NOT DONE yet\n",
    "\n",
    "## Merge VCF files\n",
    "cd /nfs/scistore18/bartogrp/apal/snap_hap/variants/jobs/postSTITCH_$run/mergeVCF\n",
    "\n",
    "for chrom in Chr{1..8}\n",
    "do \n",
    "    baseDIR=~/snap_hap/variants/$stitchRun/$chrom\n",
    "    echo -e $chrom'\\t'$baseDIR\n",
    "\n",
    "    #highConf#\n",
    "    stitchVcfList_highConf=${baseDIR}/stitchVcfs.highConf.list\n",
    "    outVCF_highConf=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.vcf.gz\n",
    "    sbatch -J postSTITCH-concatVcfs-highConf_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-concatVcfs.sbatch.sh ${chrom} ${stitchVcfList_highConf} ${outVCF_highConf}\n",
    "\n",
    "    #ALL Tagged#\n",
    "    stitchVcfList_ALL=${baseDIR}/stitchVcfs.ALL.tagged.list\n",
    "    outVCF_ALL=$baseDIR/Am_all_stitch_${chrom}_SnpGap5_biSNPs_filtered-DP_500-7732_QUAL20_MQ30.PL.tagged.vcf.gz\n",
    "    sbatch -J postSTITCH-concatVcfs-ALL_${chrom} ~/snap_hap/_scripts/sbatch/impute/job-postSTITCH-concatVcfs.sbatch.sh ${chrom} ${stitchVcfList_ALL} ${outVCF_ALL}\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cabb609-4c16-42df-8c41-dfd7d6f67c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove monomorphic SNPs to make final VCF"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
